{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Necessary imports and utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  Matrix made by matblas from blosum62.iij\n",
      "#  * column uses minimum score\n",
      "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
      "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
      "#  Cluster Percentage: >= 62\n",
      "#  Entropy =   0.6979, Expected =  -0.5209\n",
      "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
      "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
      "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0 -1.0 -4.0\n",
      "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0 -1.0 -4.0\n",
      "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0 -2.0 -4.0\n",
      "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0 -1.0 -4.0\n",
      "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -1.0 -4.0\n",
      "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0 -1.0 -4.0\n",
      "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0 -1.0 -4.0\n",
      "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0 -1.0 -4.0\n",
      "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0 -1.0 -4.0\n",
      "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0 -1.0 -4.0\n",
      "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0 -1.0 -4.0\n",
      "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0 -2.0 -4.0\n",
      "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
      "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
      "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0 -2.0 -4.0\n",
      "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0 -1.0 -4.0\n",
      "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0 -1.0 -4.0\n",
      "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "X  0.0 -1.0 -1.0 -1.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0  0.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -4.0\n",
      "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio.Align import substitution_matrices\n",
    "\n",
    "# Type hinting\n",
    "from typing import List, Set, Union, Callable\n",
    "# type sim_fct = Callable[[chr, chr], float]\n",
    "\n",
    "blossum = substitution_matrices.load(\"BLOSUM62\")\n",
    "print(blossum)\n",
    "\n",
    "\n",
    "def sim_blossum(a:chr, b:chr, gap=-8)->float:\n",
    "    if a == '-' or b == '-':\n",
    "        return gap\n",
    "    return blossum[a,b]\n",
    "\n",
    "def sim_basic(a:chr, b:chr, id=1, sub=-1, gap=-2)->float:\n",
    "    if a == b:\n",
    "        return id\n",
    "    elif a == '-' or b == '-':\n",
    "        return gap\n",
    "    else:\n",
    "        return sub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aligning two sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Needleman-Wunsch: alignment with linear gap cost\n",
    "This algorithm tries to align to protein sequences $a_1^n$ and $b_1^m$.\n",
    "By dynamic programming, the algorithm computes the maximum similarity between any pair of prefixes $a_{1...i}$ and $b_{1...j}$ and stores them in a matrix $M_{i,j}$\n",
    "\n",
    "$M_{i,j}$ is defined by recursion with the following formula:\n",
    "$$\n",
    "M_{i,j} = \\max \\begin{cases}\n",
    "  M_{i-1,j-1} + s(a_i,b_j) \\\\\n",
    "  M_{i-1,j} - \\text{gap cost} &\\\\\n",
    "  M_{i,j-1} - \\text{gap cost}&\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "To facilitate the traceback, we also compute a matrix $D$ *(like Directions)* indicating which case achieved the maximum."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_needleman_wunsch(a:str, b:str, sim_fct):\n",
    "    \"\"\" compute the Needleman-Wunsch matrix \"\"\"\n",
    "\n",
    "    # construct matrices\n",
    "    M = np.zeros((len(a)+1, len(b)+1), dtype=int)\n",
    "    D = np.zeros((len(a)+1, len(b)+1), dtype=int)\n",
    "\n",
    "    # init matrices\n",
    "    for i in range(1, len(a)+1):\n",
    "        M[i,0] = M[i-1, 0] + sim_fct(a[i-1], '-')\n",
    "    for j in range(1, len(b)+1):\n",
    "        M[0,j] = M[0, j-1] + sim_fct('-', b[j-1])\n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, len(a)+1):\n",
    "        for j in range(1, len(b)+1):\n",
    "            x1 = M[i-1,j-1]+sim_fct(a[i-1], b[j-1])\n",
    "            x2 = M[i-1,j]+sim_fct(a[i-1], '-')\n",
    "            x3 = M[i,j-1]+sim_fct('-', b[j-1])\n",
    "            if x1 >= x2 and x1 >= x3: # diagonal\n",
    "                M[i,j] = x1\n",
    "                D[i,j] = 1\n",
    "            if x2 >= x1 and x2 >= x3: # up\n",
    "                M[i,j] = x2\n",
    "                D[i,j] = 2\n",
    "            if x3 >= x1 and x3 >= x2: # left\n",
    "                M[i,j] = x3\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def trace_back(a:str, b:str, D:np.array):\n",
    "    \"\"\" traceback matrix A to get the alignment \"\"\"\n",
    "    aa = \"\"\n",
    "    bb = \"\"\n",
    "    \n",
    "    i = len(a)\n",
    "    j = len(b)\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = b[j-1] + bb\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = '-' + bb\n",
    "            i -= 1\n",
    "        else:\n",
    "            aa = '-' + aa\n",
    "            bb = b[j-1] + bb\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        aa = '-'*j + aa\n",
    "        bb = b[:j] + bb\n",
    "    elif j == 0:\n",
    "        aa = a[:i] + aa\n",
    "        bb = '-'*i + bb\n",
    "\n",
    "    return aa, bb\n",
    "\n",
    "def needleman_wunsch_global_alignment(a:str, b:str, sim_fct=sim_blossum):\n",
    "    \"\"\" Global alignment with linear gap cost \"\"\"\n",
    "    M, A = compute_matrix_needleman_wunsch(a, b, sim_fct)\n",
    "    aa, bb = trace_back(a, b, A)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return M, aa, bb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHATS\n",
      "C-AT-\n",
      "[[  0  -8 -16 -24]\n",
      " [ -8   9   1  -7]\n",
      " [-16   1   7  -1]\n",
      " [-24  -7   5   7]\n",
      " [-32 -15  -3  10]\n",
      " [-40 -23 -11   2]]\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "A = \"CHATS\"\n",
    "B = \"CAT\"\n",
    "\n",
    "F, aa, bb = needleman_wunsch_global_alignment(A, B, sim_blossum)\n",
    "print(F)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gotoh: alignment with affine gap cost\n",
    "Instead of one matrix, we need 3:\n",
    "- $F_{i,j}$ storing the best score for $a_{1...i}$ and $b_{1...j}$ that ends with $a_i$ and $b_j$\n",
    "- $A_{i,j}$ storing the best score that ends with $a_i$ and a gap\n",
    "- $B_{i,j}$ storing the best score that ends with a gap and $b_j$\n",
    "\n",
    "We have:\n",
    "$$\n",
    "A_{i,j} = \\max \\begin{cases}\n",
    "  A_{i-1,j} + \\text{gap extension} \\\\\n",
    "  F_{i-1,j} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "$$\n",
    "B_{i,j} = \\max \\begin{cases}\n",
    "  B_{i,j-1} + \\text{gap extension} \\\\\n",
    "  F_{i,j-1} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "$$\n",
    "F_{i,j} = \\max \\begin{cases}\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "The matrix $M_{i,j}$ as defined previously is the max (element wise) of this three matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_gotoh(a:str, b:str, sim_fct, gap_opening:float, gap_extension:float):\n",
    "    \"\"\" compute the Gotoh matrices \"\"\"\n",
    "    F = np.zeros((len(a)+1, len(b)+1))\n",
    "    A = np.zeros((len(a)+1, len(b)+1))\n",
    "    B = np.zeros((len(a)+1, len(b)+1))\n",
    "\n",
    "    # init matrix A\n",
    "    for j in range(0, len(b)+1):\n",
    "        A[0,j] = -np.inf\n",
    "        A[1,j] = 2*gap_opening + (j-1)*gap_extension\n",
    "    for i in range(1, len(a)+1):    \n",
    "        A[i,0] = gap_opening + (i-1)*gap_extension\n",
    "\n",
    "    # init matrix B\n",
    "    for i in range(0, len(a)+1):\n",
    "        B[i,0] = -np.inf\n",
    "        B[i,1] = 2*gap_opening + (i-1)*gap_extension\n",
    "    for j in range(1, len(b)+1):\n",
    "        B[0,j] = gap_opening + (j-1)*gap_extension\n",
    "\n",
    "    # init matrix F\n",
    "    F[0,0] = 0\n",
    "    for i in range(1, len(a)+1):\n",
    "        F[i,0] = -np.inf\n",
    "    for j in range(1, len(b)+1):\n",
    "        F[0,j] = -np.inf\n",
    "        F[0,j] = -np.inf    \n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, len(a)+1):\n",
    "        for j in range(1, len(b)+1):\n",
    "            if i>1:\n",
    "                A[i,j] = max(F[i-1,j]+gap_opening, A[i-1,j]+gap_extension)\n",
    "            if j>1:\n",
    "                B[i,j] = max(F[i,j-1]+gap_opening, B[i,j-1]+gap_extension)\n",
    "            F[i,j] = max(F[i-1,j-1]+sim_fct(a[i-1], b[j-1]), A[i-1,j-1]+sim_fct(a[i-1], b[j-1]), B[i-1,j-1]+sim_fct(a[i-1], b[j-1]))\n",
    "    \n",
    "    return F, A, B\n",
    "\n",
    "def trace_back_gotoh(a:str, b:str, F:np.array, A:np.array, B:np.array):\n",
    "    \"\"\" Does the traceback using the matrices M, A and B\n",
    "    To know in wich direction to go, we look which matrix has the highest value\"\"\"\n",
    "    aa = \"\"\n",
    "    bb = \"\"\n",
    "    i = len(a)\n",
    "    j = len(b)\n",
    "    while i>0 and j>0 :\n",
    "        m = max(F[i,j], A[i,j], B[i,j])\n",
    "        if F[i,j] == m:     # diagonal\n",
    "            aa = a[i-1] + aa\n",
    "            bb = b[j-1] + bb\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif A[i,j] == m:   # up\n",
    "            aa = a[i-1] + aa\n",
    "            bb = '-' + bb\n",
    "            i -= 1\n",
    "        else:               # left\n",
    "            aa = '-' + aa\n",
    "            bb = b[j-1] + bb\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        aa = '-'*j + aa\n",
    "        bb = b[:j] + bb\n",
    "    elif j == 0:\n",
    "        aa = a[:i] + aa\n",
    "        bb = '-'*i + bb\n",
    "\n",
    "    return aa, bb\n",
    "\n",
    "def gotoh_global_alignment(a:str, b:str, sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Global alignment with affine gap cost \"\"\"\n",
    "    F, A, B = compute_matrix_gotoh(a, b, sim_fct, gap_opening, gap_extension)\n",
    "    aa, bb = trace_back_gotoh(a, b, F, A, B)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return np.maximum(F, np.maximum(A,B)), aa, bb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHAT\n",
      "C-AT\n",
      "[[ 0. -2. -4. -6.]\n",
      " [-2.  1. -1. -3.]\n",
      " [-4. -1.  0. -2.]\n",
      " [-6. -3.  0. -1.]\n",
      " [-8. -5. -2.  1.]]\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "A = \"CHAT\"\n",
    "B = \"CAT\"\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(A, B, sim_basic, -2, -2)\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aligning N sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pairwise profile alignment\n",
    "We want to merge two alignments $A$ and $B$ into one by aligning them. For this, we can modify the Needleman-Wunsch algorithm. We just have to adapt the cost function of an alignment, to correspond to the average cost over each pair $(a,b) \\in (A,B)$.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sim_multiple(A:List[str], B:List[str], i:int, j:int, sim_fct):\n",
    "    \"\"\" Compute the average similarity between two sets of aligned sequences, at a given position \"\"\"\n",
    "    res = 0\n",
    "    for a in A:\n",
    "        for b in B:\n",
    "            res += sim_fct(a[i],b[j])\n",
    "    return res/(len(A)*len(B))\n",
    "\n",
    "def compute_matrix_needleman_wunsch_multiple(A:List[str], B:List[str], sim_fct):\n",
    "    \"\"\" Compute the Needleman-Wunsch dynamic programming matrix for two sets of aligned sequences \"\"\"\n",
    "    len_a = len(A[0])\n",
    "    len_b = len(B[0])\n",
    "\n",
    "    # construct matrices\n",
    "    M = np.zeros((len_a+1, len_b+1))\n",
    "    D = np.zeros((len_a+1, len_b+1), dtype=int)\n",
    "\n",
    "    # init matrices\n",
    "    for i in range(1, len_a+1):\n",
    "        M[i,0] = i * sim_fct('a', '-')  # cost of a gap\n",
    "    for j in range(1, len_b+1):\n",
    "        M[0,j] = j * sim_fct('a', '-')\n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, len_a+1):\n",
    "        for j in range(1, len_b+1):\n",
    "            x1 = M[i-1,j-1] + sim_multiple(A, B, i-1, j-1, sim_fct)\n",
    "            x2 = M[i-1,j] + sim_fct('a', '-')\n",
    "            x3 = M[i,j-1] + sim_fct('a', '-')\n",
    "            if x1 >= x2 and x1 >= x3: # diagonal\n",
    "                M[i,j] = x1\n",
    "                D[i,j] = 1\n",
    "            elif x2 >= x1 and x2 >= x3: # up\n",
    "                M[i,j] = x2\n",
    "                D[i,j] = 2\n",
    "            else: # x3 >= x1 and x3 >= x2: # left\n",
    "                M[i,j] = x3\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "\n",
    "def trace_back_multiple(A:List[str], B:List[str], D:np.array):\n",
    "    \"\"\" Determine the alignment given a matrix with directions \"\"\"\n",
    "\n",
    "    AA = [\"\" for _ in range(len(A))]\n",
    "    BB = [\"\" for _ in range(len(B))]\n",
    "    \n",
    "    i = len(A[0])\n",
    "    j = len(B[0])\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            for a in range(len(A)):\n",
    "                AA[a] = A[a][i-1] + AA[a]\n",
    "            for b in range(len(B)):\n",
    "                BB[b] = B[b][j-1] + BB[b]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            for a in range(len(A)):\n",
    "                AA[a] = A[a][i-1] + AA[a]\n",
    "            for b in range(len(B)):\n",
    "                BB[b] = '-' + BB[b]\n",
    "            i -= 1\n",
    "        else:\n",
    "            for a in range(len(A)):\n",
    "                AA[a] = '-' + AA[a]\n",
    "            for b in range(len(B)):\n",
    "                BB[b] = B[b][j-1] + BB[b]\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        for a in range(len(A)):\n",
    "            AA[a] = '-'*j + AA[a]\n",
    "        for b in range(len(B)):\n",
    "            BB[b] = B[b][:j] + BB[b]\n",
    "    elif j == 0:\n",
    "        for a in range(len(A)):\n",
    "            AA[a] = A[a][:i] + AA[a]\n",
    "        for b in range(len(B)):\n",
    "            BB[b] = '-'*i + BB[b]\n",
    "\n",
    "    return AA + BB   # list concatenation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0  -8 -16 -24]\n",
      " [ -8   9   1  -7]\n",
      " [-16   1   7  -1]\n",
      " [-24  -7   5   7]\n",
      " [-32 -15  -3  10]]\n",
      "\n",
      "[[  0.  -8. -16. -24.]\n",
      " [ -8.   9.   1.  -7.]\n",
      " [-16.   1.   7.  -1.]\n",
      " [-24.  -7.   5.   7.]\n",
      " [-32. -15.  -3.  10.]]\n",
      "['CHAT', 'C-AT']\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "A = \"CHAT\"\n",
    "B = \"CAT\"\n",
    "\n",
    "M, D = compute_matrix_needleman_wunsch(A, B, sim_blossum)\n",
    "print(M)\n",
    "print()\n",
    "\n",
    "M, D = compute_matrix_needleman_wunsch_multiple([A], [B], sim_blossum)\n",
    "print(M)\n",
    "\n",
    "L = trace_back_multiple([A], [B], D)\n",
    "print(L)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Progressive multiple alignment\n",
    "To align N sequences, we do clustering. First, we create one cluster per sequence. Then, we align and merge the two closest clusters until we only have one big cluster, corresponding to the alignment of all sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "\n",
    "def progressive_alignment(seq:List[str], sim_fct): # TODO: check if possible to improve efficiency\n",
    "  \"\"\" Computes multiple sequence alignement \"\"\"\n",
    "  time_mat = 0\n",
    "  time_trace = 0\n",
    "\n",
    "  # transforming each sequence into a cluster of one sequence\n",
    "  clusters = [[s] for s in seq]  \n",
    "\n",
    "  # initialising distance matrix\n",
    "  matrix = np.zeros((len(seq), len(seq)))\n",
    "  for i in range(len(seq)):\n",
    "    for j in range(i+1, len(seq)):\n",
    "      start = time()\n",
    "      matrix[i,j] = compute_matrix_needleman_wunsch(seq[i], seq[j], sim_fct)[0][-1,-1]\n",
    "      time_mat += time() - start\n",
    "\n",
    "  while len(clusters) > 1:\n",
    "    # finding closest clusters\n",
    "    max_sim = -np.inf\n",
    "    for i in range(len(clusters)):\n",
    "      for j in range(i+1,len(clusters)):\n",
    "        if matrix[i,j] > max_sim:\n",
    "          max_sim = matrix[i,j]\n",
    "          best_i = i\n",
    "          best_j = j\n",
    "\n",
    "    # merging two clusters best_i < best_j\n",
    "    a = clusters[best_i]\n",
    "    b = clusters[best_j]\n",
    "    start = time()\n",
    "    D, A = compute_matrix_needleman_wunsch_multiple(a, b, sim_fct)\n",
    "    time_mat += time() - start\n",
    "    start = time()\n",
    "    new_cluster = trace_back_multiple(a, b, A)\n",
    "    time_trace += time() - start\n",
    "\n",
    "    clusters.remove(a)\n",
    "    clusters.remove(b)\n",
    "    clusters.append(new_cluster)\n",
    "\n",
    "    # updating distance matrix\n",
    "    matrix = np.delete(matrix, [best_i,best_j], axis=0)\n",
    "    matrix = np.delete(matrix, [best_i,best_j], axis=1)\n",
    "\n",
    "    new_column = []\n",
    "    for cluster in clusters[:-1]:\n",
    "      start = time()\n",
    "      new_column.append(compute_matrix_needleman_wunsch_multiple(cluster, new_cluster, sim_fct)[0][-1,-1])\n",
    "      time_mat += time() - start\n",
    "    matrix = np.concatenate((matrix, np.array([new_column]).T), axis=1)\n",
    "    matrix = np.concatenate((matrix, np.zeros((1, len(clusters)))), axis=0)\n",
    "\n",
    "  print(\"Time for matrix computation: \", time_mat)\n",
    "  print(\"Time for trace back: \", time_trace)\n",
    "  \n",
    "  return clusters[0], D[-1,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for matrix computation:  0.0005555152893066406\n",
      "Time for trace back:  2.4557113647460938e-05\n",
      "(['-HER', 'C-AT', 'CHAT'], -10.0)\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "seq = [\"CAT\", \"CHAT\", \"HER\"]\n",
    "\n",
    "res = progressive_alignment(seq, sim_blossum)\n",
    "\n",
    "print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def read_fasta(filename):\n",
    "    \"\"\"\n",
    "    Read a fasta file and return the sequences as a list of strings\n",
    "    \"\"\"\n",
    "    sequences = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "        sequences.append(str(record.seq))\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for matrix computation:  11.2442045211792\n",
      "Time for trace back:  0.0018198490142822266\n",
      "MG-LSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAH--K---YH-\n",
      "MVHLTPEEKTAVNALWGKV--NVDAVGGEALGRLLVVYPWTQRFFESFGDLSSPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFSQLSELHCDKLHVDPENFRLLGNVLVCVLARNFGKEFTPQMQAAYQKVVAGVANALAH--K---YH-\n",
      "MVHFTAEEKAAVTSLWSKM--NVEEAGGEALGRLLVVYPWTQRFFDSFGNLSSPSAILGNPKVKAHGKKVLTSFGDAIKNMDNLKPAFAKLSELHCDKLHVDPENFKLLGNVMVIILATHFGKEFTPEVQAAWQKLVSAVAIALAH--K---YH-\n",
      "MGHFTEEDKATITSLWGKV--NVEDAGGETLGRLLVVYPWTQRFFDSFGNLSSASAIMGNPKVKAHGKKVLTSLGDAIKHLDDLKGTFAQLSELHCDKLHVDPENFKLLGNVLVTVLAIHFGKEFTPEVQASWQKMVTAVASALSS--R---YH-\n",
      "MS-LTKTERTIIVSMWAKISTQADTIGTETLERLFLSHPQTKTYFPHF-DLH-P----GSAQLRAHGSKVVAAVGDAVKSIDDIGGALSKLSELHAYILRVDPVNFKLLSHCLLVTLAARFPADFTAEAHAAWDKFLSVVSSVLTE--K---YR-\n",
      "MA-LSAEDRALVRALWKKLGSNVGVYTTEALERTFLAFPATKTYFSHL-DLS-P----GSSQVRAHGQKVADALSLAVERLDDLPHALSALSHLHACQLRVDPASFQLLGHCLLVTLARHYPGDFSPALQASLDKFLSHVISALVS--E---YR-\n",
      "MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS-H----GSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTS--K---YR-\n",
      "Score of alignment: 58.000000000000014\n"
     ]
    }
   ],
   "source": [
    "seq = read_fasta(\"data/paralogs.fasta\")\n",
    "test = progressive_alignment(seq, sim_blossum)\n",
    "for s in test[0]:\n",
    "  print(s)\n",
    "print(f\"Score of alignment: {test[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test protein with PDB: brouillon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Model id=0>\n",
      "<Chain id=A>\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB import *\n",
    "\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure(\"1MBD\", \"data/1mbd.pdb\")\n",
    "\n",
    "for model in structure:\n",
    "    print(model)\n",
    "    for chain in model:\n",
    "        print(chain)\n",
    "        for residue in chain:\n",
    "            # print(residue)\n",
    "            # print(residue.get_resname(), residue.get_id()[1], chain.id)\n",
    "            pass\n",
    "\n",
    "#dssp = DSSP(model, )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
