{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Content\n",
    "\n",
    "slides (pdf) + code\n",
    "\n",
    "similar to scientific articles:\n",
    "- context/introduction (litterature review)\n",
    "- results (take perspective, present your work as a story: problematic, progression...)\n",
    "- evaluat algos on provided dataset ( balibase)\n",
    "- discussion: critizise, propose solution, give perspective (lol), personnal thinking\n",
    "\n",
    "TODO:\n",
    "- benchmark: compare own alignments with clustalW (see scores on slides td0, or available on: https://www.genome.jp/tools-bin/clustalw) and balibase (reference); scores given by bali_score\n",
    "- implement Needlemand-Wunsch using PDB\n",
    "\n",
    "penalité de gap dans clustalW:\n",
    "10 open et 0.1 extend\n",
    "\n",
    "coef de similarité dépendant de l'enfouissement:\n",
    "- *1.5 pour burial=1\n",
    "- *0.5 pour burial=0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Necessary imports and utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  Matrix made by matblas from blosum62.iij\n",
      "#  * column uses minimum score\n",
      "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
      "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
      "#  Cluster Percentage: >= 62\n",
      "#  Entropy =   0.6979, Expected =  -0.5209\n",
      "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
      "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
      "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0 -1.0 -4.0\n",
      "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0 -1.0 -4.0\n",
      "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0 -2.0 -4.0\n",
      "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0 -1.0 -4.0\n",
      "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -1.0 -4.0\n",
      "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0 -1.0 -4.0\n",
      "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0 -1.0 -4.0\n",
      "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0 -1.0 -4.0\n",
      "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0 -1.0 -4.0\n",
      "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0 -1.0 -4.0\n",
      "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0 -1.0 -4.0\n",
      "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0 -2.0 -4.0\n",
      "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
      "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
      "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0 -2.0 -4.0\n",
      "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0 -1.0 -4.0\n",
      "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0 -1.0 -4.0\n",
      "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "X  0.0 -1.0 -1.0 -1.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0  0.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -4.0\n",
      "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.PDB import *\n",
    "import os\n",
    "\n",
    "# Type hinting\n",
    "from typing import List, Tuple, Set, Union, Callable\n",
    "# type sim_fct = Callable[[chr, chr], float]\n",
    "\n",
    "blossum = substitution_matrices.load(\"BLOSUM62\")\n",
    "print(blossum)\n",
    "\n",
    "def sim_blossum(seq_a:str, seq_b:str, i, j)->float:\n",
    "    a = seq_a[i]\n",
    "    b = seq_b[j]\n",
    "    return blossum[a,b]\n",
    "\n",
    "def sim_basic(seq_a:str, seq_b:str, i, j, id=1, sub=-1)->float:\n",
    "    a = seq_a[i]\n",
    "    b = seq_b[j]\n",
    "    if a == b:\n",
    "        return id\n",
    "    else:\n",
    "        return sub\n",
    "\n",
    "# used for profile alignment: is the average similarity between two sets of aligned sequences\n",
    "def sim_multiple(align_A:List[str], align_B:List[str], i:int, j:int, sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Compute the average similarity between two sets of aligned sequences, at a given position \"\"\"\n",
    "    res = 0\n",
    "    n = 0\n",
    "    for seq_a in align_A:\n",
    "        for seq_b in align_B:\n",
    "            if seq_a[i] == '-':\n",
    "                if i>0 and seq_a[i-1] == '-':\n",
    "                    res += gap_extension\n",
    "                else:\n",
    "                    res += gap_opening            \n",
    "            elif seq_b[j] == '-':\n",
    "                if j>0 and seq_b[j-1] == '-':\n",
    "                    res += gap_extension\n",
    "                else:\n",
    "                    res += gap_opening\n",
    "            else:\n",
    "                res += sim_fct(seq_a, seq_b, i, j)\n",
    "            n += 1\n",
    "    return res/n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(filename):\n",
    "    \"\"\" Read a fasta file and return the sequences as a list of strings \"\"\"\n",
    "    sequences = []\n",
    "    ids = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "        sequences.append(str(record.seq))\n",
    "        ids.append(record.id)\n",
    "    return sequences, ids\n",
    "\n",
    "def write_fasta(filename, sequences, ids=None, len_line=80):\n",
    "    \"\"\" Write a list of sequences to a fasta file \"\"\"\n",
    "    if ids is None:\n",
    "        ids = [f\"seq{i}\" for i in range(len(sequences))]\n",
    "    with open(filename, \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(\">{}\\n\".format(ids[i]))\n",
    "            while len(seq) > len_line:\n",
    "                f.write(seq[:len_line] + \"\\n\")\n",
    "                seq = seq[len_line:]\n",
    "            f.write(seq + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pairwise sequence alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Needleman-Wunsch: alignment with linear gap cost\n",
    "This algorithm tries to align to protein sequences $a_1^n$ and $b_1^m$.\n",
    "By dynamic programming, the algorithm computes the maximum similarity between any pair of prefixes $a_{1...i}$ and $b_{1...j}$ and stores them in a matrix $M_{i,j}$\n",
    "\n",
    "$M_{i,j}$ is defined by recursion with the following formula:\n",
    "$$\n",
    "M_{i,j} = \\max \\begin{cases}\n",
    "  M_{i-1,j-1} + s(a_i,b_j) \\\\\n",
    "  M_{i-1,j} - \\text{gap cost} &\\\\\n",
    "  M_{i,j-1} - \\text{gap cost}&\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "To facilitate the traceback, we also compute a matrix $D$ *(like Directions)* indicating which case achieved the maximum."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_needleman_wunsch(seq_a:str, seq_b:str, fct=sim_basic, gap_cost=-2):\n",
    "    \"\"\" compute the Needleman-Wunsch matrix \"\"\"\n",
    "\n",
    "    # check if we have a list of sequences (profile alignment)\n",
    "    if type(seq_a) == list:\n",
    "        n_a = len(seq_a[0])\n",
    "        n_b = len(seq_b[0])\n",
    "        sim_fct = lambda a, b, i, j: sim_multiple(a, b, i, j, fct, gap_cost, gap_cost)\n",
    "    # or a single sequence (pairwise alignment)\n",
    "    else:\n",
    "        n_a = len(seq_a)\n",
    "        n_b = len(seq_b)\n",
    "        sim_fct = fct\n",
    "\n",
    "    # construct matrices\n",
    "    M = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "    D = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "\n",
    "    # init matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        M[i,0] = M[i-1, 0] + gap_cost\n",
    "    for j in range(1, n_b+1):\n",
    "        M[0,j] = M[0, j-1] + gap_cost\n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            x1 = M[i-1,j-1] + sim_fct(seq_a, seq_b, i-1, j-1)\n",
    "            x2 = M[i-1,j] + gap_cost\n",
    "            x3 = M[i,j-1] + gap_cost\n",
    "            if x1 >= x2 and x1 >= x3: # diagonal\n",
    "                M[i,j] = x1\n",
    "                D[i,j] = 1\n",
    "            if x2 >= x1 and x2 >= x3: # up\n",
    "                M[i,j] = x2\n",
    "                D[i,j] = 2\n",
    "            if x3 >= x1 and x3 >= x2: # left\n",
    "                M[i,j] = x3\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def trace_back(a:str, b:str, D:np.array):\n",
    "    \"\"\" traceback matrix D to get the alignment \"\"\"\n",
    "    aa = \"\"\n",
    "    bb = \"\"\n",
    "    \n",
    "    i = len(a)\n",
    "    j = len(b)\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = b[j-1] + bb\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = '-' + bb\n",
    "            i -= 1\n",
    "        else:\n",
    "            aa = '-' + aa\n",
    "            bb = b[j-1] + bb\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        aa = '-'*j + aa\n",
    "        bb = b[:j] + bb\n",
    "    elif j == 0:\n",
    "        aa = a[:i] + aa\n",
    "        bb = '-'*i + bb\n",
    "\n",
    "    return aa, bb\n",
    "\n",
    "def needleman_wunsch_global_alignment(a:str, b:str, sim_fct=sim_blossum, gap_cost=-8):\n",
    "    \"\"\" Global alignment with linear gap cost \"\"\"\n",
    "    M, D = compute_matrix_needleman_wunsch(a, b, sim_fct, gap_cost)\n",
    "    aa, bb = trace_back(a, b, D)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return M, aa, bb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gotoh: alignment with affine gap cost"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of one matrix, we need 3:\n",
    "- $F_{i,j}$ storing the best score for $a_{1...i}$ and $b_{1...j}$ that ends with $a_i$ and $b_j$\n",
    "- $A_{i,j}$ storing the best score that ends with $a_i$ and a gap\n",
    "- $B_{i,j}$ storing the best score that ends with a gap and $b_j$\n",
    "\n",
    "We have:\n",
    "$$\n",
    "A_{i,j} = \\max \\begin{cases}\n",
    "  A_{i-1,j} + \\text{gap extension} \\\\\n",
    "  F_{i-1,j} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "$$\n",
    "B_{i,j} = \\max \\begin{cases}\n",
    "  B_{i,j-1} + \\text{gap extension} \\\\\n",
    "  F_{i,j-1} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "$$\n",
    "F_{i,j} = \\max \\begin{cases}\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "The matrix $M_{i,j}$ as defined previously is the max (element wise) of this three matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_gotoh(seq_a:str, seq_b:str, fct, gap_opening:float=-2, gap_extension:float=-2):\n",
    "    \"\"\" compute the Gotoh matrices \"\"\"\n",
    "\n",
    "    # check if we have a list of sequences (profile alignment)\n",
    "    if type(seq_a) == list:\n",
    "        n_a = len(seq_a[0])\n",
    "        n_b = len(seq_b[0])\n",
    "        sim_fct = lambda a, b, i, j: sim_multiple(a, b, i, j, fct, gap_opening, gap_extension)\n",
    "    # or a single sequence (pairwise alignment)\n",
    "    else:\n",
    "        n_a = len(seq_a)\n",
    "        n_b = len(seq_b)\n",
    "        sim_fct = fct\n",
    "    \n",
    "    F = np.zeros((n_a+1, n_b+1))\n",
    "    A = np.zeros((n_a+1, n_b+1))\n",
    "    B = np.zeros((n_a+1, n_b+1))\n",
    "\n",
    "    # init matrix A\n",
    "    for j in range(0, n_b+1):\n",
    "        A[0,j] = -np.inf\n",
    "        A[1,j] = 2*gap_opening + (j-1)*gap_extension\n",
    "    for i in range(1, n_a+1):    \n",
    "        A[i,0] = gap_opening + (i-1)*gap_extension\n",
    "\n",
    "    # init matrix B\n",
    "    for i in range(0, n_a+1):\n",
    "        B[i,0] = -np.inf\n",
    "        B[i,1] = 2*gap_opening + (i-1)*gap_extension\n",
    "    for j in range(1, n_b+1):\n",
    "        B[0,j] = gap_opening + (j-1)*gap_extension\n",
    "\n",
    "    # init matrix F\n",
    "    F[0,0] = 0\n",
    "    for i in range(1, n_a+1):\n",
    "        F[i,0] = -np.inf\n",
    "    for j in range(1, n_b+1):\n",
    "        F[0,j] = -np.inf\n",
    "        F[0,j] = -np.inf    \n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if i>1:\n",
    "                A[i,j] = max(F[i-1,j] + gap_opening, A[i-1,j] + gap_extension)\n",
    "            if j>1:\n",
    "                B[i,j] = max(F[i,j-1] + gap_opening, B[i,j-1] + gap_extension)\n",
    "            sim = sim_fct(seq_a, seq_b, i-1, j-1)\n",
    "            F[i,j] = max(F[i-1, j-1] + sim, A[i-1,j-1] + sim, B[i-1,j-1] + sim)\n",
    "\n",
    "    # get score and traceback matrix\n",
    "    M = np.maximum(F, np.maximum(A,B))\n",
    "    D = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if M[i,j] == F[i,j]:\n",
    "                D[i,j] = 1\n",
    "            elif M[i,j] == A[i,j]:\n",
    "                D[i,j] = 2\n",
    "            else:\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def gotoh_global_alignment(seq_a:str, seq_b:str, sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Global alignment with affine gap cost \"\"\"\n",
    "    M, D = compute_matrix_gotoh(seq_a, seq_b, sim_fct, gap_opening, gap_extension)\n",
    "    seq_aa, seq_bb = trace_back(seq_a, seq_b, D)\n",
    "    print(seq_aa)\n",
    "    print(seq_bb)\n",
    "    return M, seq_aa, seq_bb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHATSASA\n",
      "C-AT-A-A\n",
      "[[  0  -8 -16 -24 -32 -40]\n",
      " [ -8   9   1  -7 -15 -23]\n",
      " [-16   1   7  -1  -9 -17]\n",
      " [-24  -7   5   7   3  -5]\n",
      " [-32 -15  -3  10   7   3]\n",
      " [-40 -23 -11   2  11   8]\n",
      " [-48 -31 -19  -6   6  15]\n",
      " [-56 -39 -27 -14  -2   7]\n",
      " [-64 -47 -35 -22 -10   2]]\n",
      "CHATSASA\n",
      "C-AT-A-A\n",
      "[[  0.  -8. -16. -24. -32. -40.]\n",
      " [ -8.   9.   1.  -7. -15. -23.]\n",
      " [-16.   1.   7.  -1.  -9. -17.]\n",
      " [-24.  -7.   5.   7.   3.  -5.]\n",
      " [-32. -15.  -3.  10.   7.   3.]\n",
      " [-40. -23. -11.   2.  11.   8.]\n",
      " [-48. -31. -19.  -6.   6.  15.]\n",
      " [-56. -39. -27. -14.  -2.   7.]\n",
      " [-64. -47. -35. -22. -10.   2.]]\n",
      "CHATSASA\n",
      "C-AT--AA\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "align_a = \"CHATSASA\"\n",
    "align_b = \"CATAA\"\n",
    "\n",
    "M, aa, bb = needleman_wunsch_global_alignment(align_a, align_b, sim_blossum, -8)\n",
    "print(M)\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_blossum, -8, -8)\n",
    "print(M)\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_basic, -5, -2)\n",
    "# print(M)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pairwise profile alignment\n",
    "We want to merge two alignments $A$ and $B$ into one by aligning them. To compute the similarity matrix, we just have to modify the previous algorithms by replacing the similarity cost of two bases by the average similarity over all the bases in the two alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trace_back_multiple(align_a:List[str], align_b:List[str], D:np.array):\n",
    "    \"\"\" Determine the alignment given a matrix with directions \"\"\"\n",
    "\n",
    "    align_aa = [\"\" for _ in range(len(align_a))]\n",
    "    align_bb = [\"\" for _ in range(len(align_b))]\n",
    "    \n",
    "    i = len(align_a[0])\n",
    "    j = len(align_b[0])\n",
    "    \n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = align_a[a][i-1] + align_aa[a]\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = align_b[b][j-1] + align_bb[b]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = align_a[a][i-1] + align_aa[a]\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = '-' + align_bb[b]\n",
    "            i -= 1\n",
    "        else:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = '-' + align_aa[a]\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = align_b[b][j-1] + align_bb[b]\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        for a in range(len(align_a)):\n",
    "            align_aa[a] = '-'*j + align_aa[a]\n",
    "        for b in range(len(align_b)):\n",
    "            align_bb[b] = align_b[b][:j] + align_bb[b]\n",
    "    elif j == 0:\n",
    "        for a in range(len(align_a)):\n",
    "            align_aa[a] = align_a[a][:i] + align_aa[a]\n",
    "        for b in range(len(align_b)):\n",
    "            align_bb[b] = '-'*i + align_bb[b]\n",
    "\n",
    "    return align_aa + align_bb   # list concatenation\n",
    "\n",
    "def needleman_wunsch_global_alignment_multiple(align_a:List[str], align_b:List[str], sim_fct=sim_basic, gap_cost=-2):\n",
    "    \"\"\" Global alignment with linear gap cost \"\"\"\n",
    "    M, D = compute_matrix_needleman_wunsch(align_a, align_b, sim_fct, gap_cost)\n",
    "    aa, bb = trace_back_multiple(align_a, align_b, D)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return M, aa, bb\n",
    "\n",
    "def gotoh_global_alignment_multiple(align_a, align_b, sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Global alignment with affine gap cost \"\"\"\n",
    "    M, D = compute_matrix_gotoh(align_a, align_b, sim_fct, gap_opening, gap_extension)\n",
    "    aa, bb = trace_back_multiple(align_a, align_b, D)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return M, aa, bb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHAT\n",
      "C-AT\n",
      "CHAT\n",
      "C-AT\n",
      "CHAT\n",
      "C-AT\n",
      "CHAT\n",
      "C-AT\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "align_a = \"CHAT\"\n",
    "align_b = \"CAT\"\n",
    "\n",
    "M, aa, bb = needleman_wunsch_global_alignment(align_a, align_b, sim_basic, -2)\n",
    "\n",
    "M, aa, bb = needleman_wunsch_global_alignment_multiple([align_a], [align_b], sim_basic, -2)\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_basic, -2)\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment_multiple([align_a], [align_b], sim_basic, -2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Progressive multiple alignment\n",
    "To align N sequences, we do clustering. First, we create one cluster per sequence. Then, we align and merge the two closest clusters until we only have one big cluster, containing all the alignment sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def progressive_alignment(seq:list, sim_fct=sim_blossum, gap_opening=-10, gap_extension=-0.5, linear=False):\n",
    "  \"\"\" Computes multiple sequence alignement \"\"\"\n",
    "\n",
    "  # transforming each sequence into a cluster of one sequence\n",
    "  clusters = [[s] for s in seq]\n",
    "  # order of the clusters, to keep track of the original order\n",
    "  order = list(range(len(seq)))\n",
    "\n",
    "  # initialising distance matrix\n",
    "  matrix = np.zeros((len(seq), len(seq)))\n",
    "  for i in range(len(seq)):\n",
    "    for j in range(i+1, len(seq)):\n",
    "      if linear:\n",
    "        matrix[i,j] = compute_matrix_needleman_wunsch(seq[i], seq[j], sim_fct, gap_opening)[0][-1,-1]\n",
    "      else:\n",
    "        matrix[i,j] = compute_matrix_gotoh(seq[i], seq[j], sim_fct, gap_opening, gap_extension)[0][-1,-1]\n",
    "      matrix[j,i] = matrix[i,j]\n",
    "\n",
    "  while len(clusters) > 1:\n",
    "    # finding closest clusters (cad with highest similarity)\n",
    "    max_sim = -np.inf\n",
    "    for i in range(len(clusters)):\n",
    "      for j in range(i+1,len(clusters)):\n",
    "        if matrix[i,j] > max_sim:\n",
    "          max_sim = matrix[i,j]\n",
    "          best_i = i\n",
    "          best_j = j\n",
    "\n",
    "    # merging two clusters best_i < best_j\n",
    "    cluster_j = clusters.pop(best_j)\n",
    "    cluster_i = clusters.pop(best_i)\n",
    "    \n",
    "    order_j = order.pop(best_j)\n",
    "    order_i = order.pop(best_i)\n",
    "    order.append([order_i, order_j])\n",
    "\n",
    "    if linear:\n",
    "      M, D = compute_matrix_needleman_wunsch(cluster_i, cluster_j, sim_fct, gap_opening)\n",
    "    else:\n",
    "      M, D = compute_matrix_gotoh(cluster_i, cluster_j, sim_fct, gap_opening, gap_extension)\n",
    "    new_cluster = trace_back_multiple(cluster_i, cluster_j, D)\n",
    "\n",
    "    clusters.append(new_cluster)\n",
    "\n",
    "    # updating distance matrix\n",
    "    new_line = (len(cluster_i)*matrix[best_i,:] + len(cluster_j)*matrix[best_j,:]) / (len(cluster_i)+len(cluster_j))  # weighted average\n",
    "    new_column = np.append(new_line, 0)  # adding a column of 0\n",
    "    new_line = new_line.reshape(1, -1)\n",
    "    new_column = new_column.reshape(-1, 1)\n",
    "\n",
    "    matrix = np.append(matrix, new_line, axis=0)\n",
    "    matrix = np.append(matrix, new_column, axis=1)\n",
    "\n",
    "    matrix = np.delete(matrix, [best_i, best_j], axis=0)\n",
    "    matrix = np.delete(matrix, [best_i, best_j], axis=1)\n",
    "\n",
    "  def get_order(gt):\n",
    "    \"\"\"((4, (0, 1)), (2, 3)) -> [4, 0, 1, 2, 3]\"\"\"\n",
    "    if type(gt) == int:\n",
    "        return [gt]\n",
    "    else:\n",
    "        return get_order(gt[0]) + get_order(gt[1])\n",
    "    \n",
    "  order = get_order(order[0])\n",
    "  reverse_order = [order.index(i) for i in range(len(order))]\n",
    "  \n",
    "  alignment = clusters[0]\n",
    "  alignment = [alignment[i] for i in reverse_order]\n",
    "  \n",
    "  return alignment, D[-1,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C-AT\n",
      "CHAT\n",
      "-HER\n",
      "CATS\n"
     ]
    }
   ],
   "source": [
    "# test: easy example\n",
    "seq = [\"CAT\", \"CHAT\", \"HER\", \"CATS\"]\n",
    "\n",
    "align, score = progressive_alignment(seq, sim_blossum)\n",
    "\n",
    "for seq in align:\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SISDTVKRAREAFNSGKTRSLQFRIQQLEALQRMINENLKSISGALASDLGKNEWTSYYEEVAHVLEELDTTIKELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLTIQPMVGAVAAGNAVILKPSEVSGHMADLLATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTPVTLELGGKSPCYVDKDCDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRDYGRIINDRHFQRVKGLIDNQKVAHGGTWDQSSRYIAPTILVDVDPQSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNEKVIKKMIAETSSGGVTANDVIVHITVPTLPFGGVGNSGMGAYHGKKSFETFSHRRSCLVKSLLNEEAHKARYPPSPA\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEATLVYEVGKNWVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGEDNESFYVPLGAGVVIAPWNFPVAIFTGMIVGPVAVGNTVIAKPAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYLVEHPRIRFINFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIVDETADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPDLGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVFTEVPPKARIAQEEIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKREHLEWARREFHVGNLYFNRKITGALVGVQPFGGFKLSGTNAKTGALDYLRLFLEMKAVAERF\n",
      "TDNVFYATNAFTGEALPLAFPVHTEVEVNQAATAAAKVARDFRRLNNSKRASLLRTIASELEARSDDIIARAHLETALPEVRLTGEIARTANQLRLFADVVNSGSYHQAILDTPNPTRAPLPKPDIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGTSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQALVSHPEIKAVGFTGSVGGGRALFNLAHERPEPIPFYGELGAINPTFIFPSAMRAKADLADQFVASMTMGCGQFCTKPGVVFALNTPETQAFIETAQSLIRQQSPSTLLTPGIRDSYQSQVVSRGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "DELLEKAKKVREAWDVLRNATTREKNKAIKKIAEKLDERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDEXIKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGIIYESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSAIREALKETEIPESSVEFIENTDRSLVLEXIRLREYLSLVIPRGGYGLISFVRDNATVPVLETGVGNCHIFVDESADLKKAVPVIINAKTQRPGTCNAAEKLLVHEKIAKEFLPVIVEELRKHGVEVRGCEKTREIVPDVVPATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESILTENYSNAKKFVSEIDAAAVYVNASTRFTDGGQFGFGAEIGISTQRFHARGPVGLRELTTYKFVVLGEYHVRE\n",
      "\n",
      "---------------------SISDTVKRAREAF--------------------------------NSGKT-RSLQFRIQQLEALQRMI----N---EN---------------LKS-ISGALASDLGKNEWTSYYEEVAHVLE--ELD----TTIK----ELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLT-I-QPMVGAVAAGNAVILK-----PSE--------VSG--HMADL---LATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGST-AVGKIVMAAAAKHLTP------VTLELGGKSPCYV--DKD-CDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRD--YGRIINDRHFQRVKG-L-I---DNQKVAHG--G---T-WDQSSRYIAPTIL-VDVDP--QSPVMQ-E---EIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNE---KVIKKM-IAETSSGGVTAN----D-VIVHITV--PTLPFGG--VGNS-GMGAYHGKKSFETFSH--RRSCLVKSLLNEEAHK-AR-YPPSPA\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKA-GKAEA-EAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMR-RRKRELEATLVYEVGKN-WVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGED---------NESFYV---PLGAGVVIAPWNFPVAIF-T-GMIVGPVAVGNTVIAK-----PAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYL----V----EHPRIRF--INFTGSL-EVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIV--DET-ADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPD--LGPVVSAEQERKVLSYIEIGKNEGQLVLGG--K---R-LEGEGYFIAPTVF-TEVPP--KARIAQ-E---EIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKR---EHLEWA-RREFHVGNLYFN----R-KITGALV--GVQPFGG--FKLS-GTNA----KTG-ALDY-------LRLFLEMKAVA-ER-F-----\n",
      "-----------------------TDNVFYATNAF--------------------------TGEALPLAFPVHTEVEVNQAAT-AAAKVARDFRRLNNSKRASLLRTIASELE-ARSDDIIARAHLETALPE-VRLTGEIARTANQLRLFADVVNSGSYH-QAILDTPNPTRAPLPKPDIRRQQ----IALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGT-SQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQAL--------VSHPEIK--AVGFTGSV-GGGRALFNLAHERPEP----IPFYGELGAINPTFIFPSAMRAKADLADQFVA-SMTMGCGQFCTKPG-VVF--ALNTPETQAFIETAQSL-I--RQQSPSTLLTPGIRDSYQSQVVS------------RGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWE-E---EIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALP-ESLLPDSLKAENPLEIARAVDGKAA\n",
      "--------------------DELLEKAKKVREA------------W----------------------DVL-RNATT-REKNKAIKKIAEKL----DERRKEILEANRIDVEKARERGVKESLVDRLALN-----DKRIDEXIKACETVIGLKD----PVGEVIDSWVRED------GLRIAR-V-RVPIGPIGII--YESRPNVT-V-ETTILALKSGNTILLRGGSDALNS-NKAIVSAIREALKETEIPESSVEFIENT-DRS---L-------VLEXIRLR---EYLSLVIPRGGYGLISFVRDNA---T---VPVLETGVGN-CHIFVDES-ADLKKAVPVIINAKTQRPG-TCNA---------A-----EKLLV-HEKI----AK---E-FL-P-VI---VEE-LR-----K---HGVEV--RG-----C-EKTREIVP-----DVV-----PATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESIL-TEN-YSNAKKFV--SEIDAAAVYVN-------ASTRFTDGGQFGFG----A-EIGISTQRFHARG-PV-G-L-RELTTYKFVVLGEYHV-R------E\n",
      "Score of alignment: 1\n",
      "Size of alignment: 601\n"
     ]
    }
   ],
   "source": [
    "# test: real example\n",
    "seq, ids = read_fasta(\"data/balibase/RV11.unaligned/BB11003.fasta\")\n",
    "#seq = read_fasta(\"data/paralogs.fasta\")\n",
    "\n",
    "for s in seq:\n",
    "  print(s)\n",
    "print()\n",
    "\n",
    "alignment, score = progressive_alignment(seq, sim_blossum, gap_opening=-10, gap_extension=-0.5, linear=False)\n",
    "for s in alignment:\n",
    "  print(s)\n",
    "print(f\"Score of alignment: {score}\")\n",
    "print(f\"Size of alignment: {len(alignment[0])}\")\n",
    "\n",
    "filename = \"data/balibase/results/BB11003.fasta\"\n",
    "write_fasta(filename, alignment, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Include structural information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and read PDB files\n",
    "DSSP keys:\n",
    "indice de chaine, indice de residue (avec code d'insertion, souvent vide)\n",
    "\n",
    "indice de residu, type de residu, structure secondaire (E if beta strain, H, other), solvant accessibility, phi, psi, other..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dssp(mol:str):\n",
    "    # download PDB file\n",
    "    PDBfile = f\"data/pdb/{mol}.pdb\"\n",
    "    if not os.path.exists(PDBfile):\n",
    "        pdbl = PDBList()\n",
    "        oldPDBfile = pdbl.retrieve_pdb_file(mol, pdir='data/pdb/', file_format=\"pdb\")\n",
    "        os.rename(oldPDBfile, PDBfile)\n",
    "\n",
    "    # parse PDB file\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure(mol, PDBfile)\n",
    "\n",
    "    # get DSSP\n",
    "    model = structure[0]\n",
    "    dssp = DSSP(model, PDBfile)\n",
    "    return model, dssp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 'M', '-', 1.0, 360.0, -66.0, 0, 0.0, 0, 0.0, 0, 0.0, 0, 0.0)\n",
      "(2, 'E', '-', 0.7628865979381443, 64.4, 100.8, 2, -0.1, 3, -0.0, 1, -0.0, 0, 0.0)\n",
      "(3, 'V', '-', 0.7394366197183099, -62.9, 147.7, 1, -0.1, 2, -1.6, 0, 0.0, -1, -0.0)\n",
      "(4, 'K', '-', 0.9414634146341463, -65.1, 89.4, 2, -0.0, 2, -0.2, 3, -0.0, -1, -0.1)\n",
      "(5, 'K', '-', 0.4926829268292683, -89.1, 156.7, -2, -1.6, 2, -1.9, 1, -0.1, -1, -0.0)\n",
      "(6, 'T', '-', 0.6197183098591549, -70.5, 84.3, -2, -0.2, 2, -1.3, 1, -0.2, -1, -0.1)\n"
     ]
    }
   ],
   "source": [
    "# test        \n",
    "mol = \"1idy\"\n",
    "model, dssp = get_dssp(mol)\n",
    "\n",
    "for chain in model:\n",
    "    i = 0\n",
    "    for residue in chain:\n",
    "        print(dssp[(chain.id, residue.id)])\n",
    "        i += 1\n",
    "        if i > 5:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta_and_PDB(filename) -> Tuple[List[str], List[List[Tuple]]]:\n",
    "    \"\"\"\n",
    "    Read a fasta file and the corresponding PDB files\n",
    "    Returns:\n",
    "    - seqFASTA, list of str: the sequences in the fasta file\n",
    "    - DSSPs, list of list of tuple: each tuple corresponds to the DSSP data of a residue (index, name, secondary structure, burial, ... )\n",
    "    \"\"\"\n",
    "    seqFASTA = []\n",
    "    DSSPs = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "\n",
    "        seqFASTA.append(str(record.seq))\n",
    "\n",
    "        name = record.name\n",
    "        mol, chain = name.split(\"_\")\n",
    "        if chain == \"\":\n",
    "            chain = \"A\"\n",
    "        model, dssp = get_dssp(mol)\n",
    "\n",
    "        DSSPextract = []\n",
    "        for residue in model[chain]:\n",
    "            if is_aa(residue):\n",
    "                try:\n",
    "                    DSSPextract.append(dssp[(chain, residue.id)])\n",
    "                except KeyError:\n",
    "                    print(f\"residue {residue} not found in DSSP for {mol}_{chain}\")\n",
    "        DSSPs.append(DSSPextract)\n",
    "    return seqFASTA, DSSPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align fasta and PDB\n",
    "Some amino-acids might be missing in the PDB file. Therefore we have to use Needleman-Wunsch to align properly the sequence obtained from the fasta and from the PDB file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trace_back_DSSP(seq:str, dssp:List[tuple], D:np.array) -> Tuple[str, List[tuple]]:\n",
    "    \"\"\" traceback matrix D to align a sequence and a list with the data from the PDB file \"\"\"\n",
    "    new_seq = \"\"\n",
    "    new_dssp = []\n",
    "    \n",
    "    i = len(seq)\n",
    "    j = len(dssp)\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            new_seq += seq[i-1]\n",
    "            new_dssp.append(dssp[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            new_seq += seq[i-1]\n",
    "            new_dssp.append(tuple(['-' for _ in range(14)]))\n",
    "            i -= 1\n",
    "        else:\n",
    "            new_seq += '-'\n",
    "            new_dssp.append(dssp[j-1])\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        new_seq += '-'*j\n",
    "        new_dssp += dssp[:j]\n",
    "    elif j == 0:\n",
    "        new_seq += seq[:i]\n",
    "        new_dssp += [tuple() for _ in range(i)]     \n",
    "    return new_seq[::-1], new_dssp[::-1]\n",
    "\n",
    "def align_DSSP(seq:str, dssp:List[tuple]):\n",
    "    \"\"\" align dssp data with brute sequence\"\"\"\n",
    "    sim_bool = lambda a, b, i, j: 1 if a[i] == b[j] else 0\n",
    "    M, D = compute_matrix_needleman_wunsch(seq, \"\".join([d[1] for d in dssp]), sim_bool, gap_cost=-1)\n",
    "    new_seq, new_dssp = trace_back_DSSP(seq, dssp, D)\n",
    "    return new_seq, new_dssp\n",
    "\n",
    "def read_and_align_fasta_and_PDB(filename) -> Tuple[List[str], List[List[Tuple]]]:\n",
    "    Lseq, Lddps = read_fasta_and_PDB(filename)\n",
    "    new_Lseq, new_Ldssp = [], []\n",
    "    for i in range(len(Lseq)):\n",
    "        seq, dssp = align_DSSP(Lseq[i], Lddps[i])\n",
    "        new_Lseq.append(seq)\n",
    "        new_Ldssp.append(dssp)\n",
    "    return new_Lseq, new_Ldssp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9025.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9069.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9113.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9240.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8685.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8701.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8733.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9485.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15917.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15948.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15979.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16010.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 16041.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 16142.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 16238.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16349.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue MSE het=H_MSE resseq=78 icode= > not found in DSSP for 1o20_A\n",
      "residue <Residue MSE het=H_MSE resseq=189 icode= > not found in DSSP for 1o20_A\n",
      "SISDTVKRAREAFNSGKTRSLQFRIQQLEALQRMINENLKSISGALASDLGKNEWTSYYEEVAHVLEELDTTIKELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLTIQPMVGAVAAGNAVILKPSEVSGHMADLLATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTPVTLELGGKSPCYVDKDCDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRDYGRIINDRHFQRVKGLIDNQKVAHGGTWDQSSRYIAPTILVDVDPQSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNEKVIKKMIAETSSGGVTANDVIVHITVPTLPFGGVGNSGMGAYHGKKSFETFSHRRSCLVKSLLNEEAHKARYPPSPA\n",
      "SISDTVKRAREAFNSGKTRSLQFRIQQLEALQRMINENLKSISGALASDLGKNEWTSYYEEVAHVLEELDTTIKELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLTIQPMVGAVAAGNAVILKPSEVSGHMADLLATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTPVTLELGGKSPCYVDKDCDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRDYGRIINDRHFQRVKGLIDNQKVAHGGTWDQSSRYIAPTILVDVDPQSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNEKVIKKMIAETSSGGVTANDVIVHITVPTLPFGGVGNSGMGAYHGKKSFETFSHRRSCLVKSLLNEEAHKARYPPSPA\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEATLVYEVGKNWVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGEDNESFYVPLGAGVVIAPWNFPVAIFTGMIVGPVAVGNTVIAKPAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYLVEHPRIRFINFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIVDETADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPDLGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVFTEVPPKARIAQEEIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKREHLEWARREFHVGNLYFNRKITGALVGVQPFGGFKLSGTNAKTGALDYLRLFLEMKAVAERF\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEATLVYEVGKNWVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGEDNESFYVPLGAGVVIAPWNFPVAIFTGMIVGPVAVGNTVIAKPAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYLVEHPRIRFINFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIVDETADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPDLGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVFTEVPPKARIAQEEIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKREHLEWARREFHVGNLYFNRKITGALVGVQPFGGFKLSGTNAKTGALDYLRLFLEMKAVAERF\n",
      "TDNVFYATNAFTGEALPLAFPVHTEVEVNQAATAAAKVARDFRRLNNSKRASLLRTIASELEARSDDIIARAHLETALPEVRLTGEIARTANQLRLFADVVNSGSYHQAILDTPNPTRAPLPKPDIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGTSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQALVSHPEIKAVGFTGSVGGGRALFNLAHERPEPIPFYGELGAINPTFIFPSAMRAKADLADQFVASMTMGCGQFCTKPGVVFALNTPETQAFIETAQSLIRQQSPSTLLTPGIRDSYQSQVVSRGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "TDNVFYATNAFTGEALPLAFPVHTEVEVNQAATAAAKVARDFRRLNNSKRASLLRTIASELEARSDDIIARAHLETALPEVRLTGEIARTANQLRLFADVVNSGSYHQAILDTPNPTRAPLPKPDIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGTSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQALVSHPEIKAVGFTGSVGGGRALFNLAHERPEPIPFYGELGAINPTFIFPSAMRAKADLADQFVASMTMGCGQFCTKPGVVFALNTPETQAFIETAQSLIRQQSPSTLLTPGIRDSYQSQVVSRGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "DELLEKAKKVREAWDVLRNATTREKNKAIKKIAEKLDERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDEXIKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGIIYESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSAIREALKETEIPESSVEFIENTDRSLVLEXIRLREYLSLVIPRGGYGLISFVRDNATVPVLETGVGNCHIFVDESADLKKAVPVIINAKTQRPGTCNAAEKLLVHEKIAKEFLPVIVEELRKHGVEVRGCEKTREIVPDVVPATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESILTENYSNAKKFVSEIDAAAVYVNASTRFTDGGQFGFGAEIGISTQRFHARGPVGLRELTTYKFVVLGEYHVRE\n",
      "DELLEKAKKVREAWDVLRNATTREKNKAIKKIAEKLDERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDE-IKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGIIYESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSAIREALKETEIPESSVEFIENTDRSLVLE-IRLREYLSLVIPRGGYGLISFVRDNATVPVLETGVGNCHIFVDESADLKKAVPVIINAKTQRPGTCNAAEKLLVHEKIAKEFLPVIVEELRKHGVEVRGCEKTREIVPDVVPATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESILTENYSNAKKFVSEIDAAAVYVNASTRFTDGGQFGFGAEIGISTQRFHARGPVGLRELTTYKFVVLGEYHVRE\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "seqFASTA, Dssps = read_and_align_fasta_and_PDB(\"data/balibase/RV11.unaligned/BB11003.fasta\")\n",
    "\n",
    "seqPDB = []\n",
    "burials = []\n",
    "for dssp in Dssps:\n",
    "    seqPDB.append(\"\".join([d[1] for d in dssp]))\n",
    "    burials.append([d[3] for d in dssp])\n",
    "\n",
    "for i in range(len(seqFASTA)):\n",
    "    print(seqFASTA[i])\n",
    "    print(seqPDB[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use burial to modulate similarity function in Needleman-Wunsh\n",
    "Idea: we keep the same algorithm but modify the gap or mutation cost depending of the structural information.\n",
    "\n",
    "The most easy data to use is the burial score : a mutation inside of the protein is more costly than on the outside. Therefore we use the same algorythm as previopusly but multiply the similarity by a coefficient depending on the burial.\n",
    "\n",
    "\n",
    "$M_{i,j}$ is defined by recursion with the following formula:\n",
    "$$\n",
    "M_{i,j} = \\max \\begin{cases}\n",
    "  M_{i-1,j-1} + f(x) \\cdot s(a_i,b_j) \\\\\n",
    "  M_{i-1,j} - f(x) \\cdot \\text{gap cost} &\\\\\n",
    "  M_{i,j-1} - f(x) \\cdot \\text{gap cost}&\n",
    "\\end{cases}\n",
    "$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sim_burial(dssp_a, dssp_b, i, j, gap=-8)->float:\n",
    "    ''' a, b tuples (sequence, list of burial values)'''\n",
    "    print(dssp_a)\n",
    "    bur = 0.5 + (dssp_a[i][3] + dssp_b[j][3])/2\n",
    "    return bur * sim_blossum([dssp_a[i][1]], [dssp_b[j][1]], 0, 0, gap)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Benchmark !!!\n",
    "For the sequence from balivase, we want to compare:\n",
    "- our computed alignment (progressive alignment using affine/linear gap with/without structural information)\n",
    "- the prediction from clustalW (best alignment without structural info)\n",
    "- the consensus alignment given in the fasta file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'test' from 'bali_score2' (/home/flopau/Documents/3A/P2/INFBIO588_projet_bioinfo/project/aligning_proteins/bali_score2.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[52], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbali_score2\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m  test\n\u001b[1;32m      3\u001b[0m test()\n\u001b[1;32m      5\u001b[0m file \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBB11003.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'test' from 'bali_score2' (/home/flopau/Documents/3A/P2/INFBIO588_projet_bioinfo/project/aligning_proteins/bali_score2.py)"
     ]
    }
   ],
   "source": [
    "from bali_score2 import  compare_seq, read_fasta2dict, test\n",
    "\n",
    "test()\n",
    "\n",
    "file = \"BB11003.fasta\"\n",
    "\n",
    "unaligned = \"data/balibase/RV11.unaligned/\" + file\n",
    "test = \"data/balibase/results/\" + file\n",
    "reference = \"data/balibase/RV11.aligned/\" + file\n",
    "\n",
    "seqs = read_fasta(unaligned)\n",
    "alignment, score = progressive_alignment(seq, sim_blossum, gap_opening=-10, gap_extension=-0.5, linear=False)\n",
    "write_fasta(test, alignment)\n",
    "\n",
    "refkeylist, refseq = read_fasta2dict(reference)\n",
    "testkeylist, testseq = read_fasta2dict(test)\n",
    "\n",
    "sp, tc = compare_seq(refseq, refkeylist, testseq, testkeylist)\n",
    "print(sp, tc)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
