{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notes\n",
    "\n",
    "slides (pdf) + code\n",
    "\n",
    "similar to scientific articles:\n",
    "- context/introduction (litterature review)\n",
    "- results (take perspective, present your work as a story: problematic, progression...)\n",
    "- evaluat algos on provided dataset ( balibase)\n",
    "- discussion: critizise, propose solution, give perspective (lol), personnal thinking\n",
    "\n",
    "TODO:\n",
    "- benchmark: compare own alignments with clustalW (see scores on slides td0, or available on: https://www.genome.jp/tools-bin/clustalw) and balibase (reference); scores given by bali_score\n",
    "- implement Needlemand-Wunsch using PDB\n",
    "\n",
    "penalité de gap dans clustalW:\n",
    "10 open et 0.1 extend\n",
    "\n",
    "coef de similarité dépendant de l'enfouissement:\n",
    "- *1.5 pour burial=1\n",
    "- *0.5 pour burial=0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Necessary imports and utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  Matrix made by matblas from blosum62.iij\n",
      "#  * column uses minimum score\n",
      "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
      "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
      "#  Cluster Percentage: >= 62\n",
      "#  Entropy =   0.6979, Expected =  -0.5209\n",
      "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
      "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
      "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0 -1.0 -4.0\n",
      "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0 -1.0 -4.0\n",
      "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0 -2.0 -4.0\n",
      "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0 -1.0 -4.0\n",
      "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -1.0 -4.0\n",
      "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0 -1.0 -4.0\n",
      "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0 -1.0 -4.0\n",
      "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0 -1.0 -4.0\n",
      "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0 -1.0 -4.0\n",
      "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0 -1.0 -4.0\n",
      "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0 -1.0 -4.0\n",
      "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0 -2.0 -4.0\n",
      "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
      "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
      "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0 -2.0 -4.0\n",
      "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0 -1.0 -4.0\n",
      "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0 -1.0 -4.0\n",
      "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "X  0.0 -1.0 -1.0 -1.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0  0.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -4.0\n",
      "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import warnings\n",
    "from Bio import SeqIO\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.PDB import *\n",
    "from bali_score import  compare_seq, read_fasta2dict\n",
    "\n",
    "# Type hinting\n",
    "from typing import List, Tuple, Set, Union, Callable\n",
    "\n",
    "blossum = substitution_matrices.load(\"BLOSUM62\")\n",
    "print(blossum)\n",
    "\n",
    "def sim_blossum(a:chr, b:chr)->float:\n",
    "    return blossum[a,b]\n",
    "\n",
    "def sim_basic(a:chr, b:chr, id=1, sub=-1)->float:\n",
    "    if a == b:\n",
    "        return id\n",
    "    else:\n",
    "        return sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(filename):\n",
    "    \"\"\" Read a fasta file and return the sequences as a list of strings \"\"\"\n",
    "    sequences = []\n",
    "    ids = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "        sequences.append(str(record.seq))\n",
    "        ids.append(record.id)\n",
    "    return sequences, ids\n",
    "\n",
    "def write_fasta(filename, sequences, ids=None, len_line=80):\n",
    "    \"\"\" Write a list of sequences to a fasta file \"\"\"\n",
    "    if ids is None:\n",
    "        ids = [f\"seq{i}\" for i in range(len(sequences))]\n",
    "    os.makedirs(os.path.dirname(filename), exist_ok=True)\n",
    "    with open(filename, \"w\") as f:\n",
    "        for i, seq in enumerate(sequences):\n",
    "            f.write(\">{}\\n\".format(ids[i]))\n",
    "            while len(seq) > len_line:\n",
    "                f.write(seq[:len_line] + \"\\n\")\n",
    "                seq = seq[len_line:]\n",
    "            f.write(seq + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pairwise sequence alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Needleman-Wunsch: alignment with linear gap cost\n",
    "This algorithm tries to align to protein sequences $a_1^n$ and $b_1^m$.\n",
    "By dynamic programming, the algorithm computes the maximum similarity between any pair of prefixes $a_{1...i}$ and $b_{1...j}$ and stores them in a matrix $M_{i,j}$\n",
    "\n",
    "$M_{i,j}$ is defined by recursion with the following formula:\n",
    "$$\n",
    "M_{i,j} = \\max \\begin{cases}\n",
    "  M_{i-1,j-1} + s(a_i,b_j) \\\\\n",
    "  M_{i-1,j} - \\text{gap cost} &\\\\\n",
    "  M_{i,j-1} - \\text{gap cost}&\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "To facilitate the traceback, we also compute a matrix $D$ *(like Directions)* indicating which case achieved the maximum."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_needleman_wunsch(seq_a:str, seq_b:str, sim_fct=sim_basic, gap_cost=-2):\n",
    "    \"\"\" compute the Needleman-Wunsch matrix \"\"\"\n",
    "\n",
    "    n_a = len(seq_a)\n",
    "    n_b = len(seq_b)\n",
    "\n",
    "    # construct matrices\n",
    "    M = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "    D = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "\n",
    "    # init matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        M[i,0] = M[i-1, 0] + gap_cost\n",
    "    for j in range(1, n_b+1):\n",
    "        M[0,j] = M[0, j-1] + gap_cost\n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if seq_a[i-1] == '-' or seq_b[j-1] == '-':\n",
    "                sim = gap_cost\n",
    "            else:\n",
    "                sim = sim_fct(seq_a[i-1], seq_b[j-1])\n",
    "            x1 = M[i-1,j-1] + sim\n",
    "            x2 = M[i-1,j] + gap_cost\n",
    "            x3 = M[i,j-1] + gap_cost\n",
    "            if x1 >= x2 and x1 >= x3: # diagonal\n",
    "                M[i,j] = x1\n",
    "                D[i,j] = 1\n",
    "            if x2 >= x1 and x2 >= x3: # up\n",
    "                M[i,j] = x2\n",
    "                D[i,j] = 2\n",
    "            if x3 >= x1 and x3 >= x2: # left\n",
    "                M[i,j] = x3\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def trace_back(a:str, b:str, D:np.array):\n",
    "    \"\"\" traceback matrix D to get the alignment \"\"\"\n",
    "    aa = \"\"\n",
    "    bb = \"\"\n",
    "    \n",
    "    i = len(a)\n",
    "    j = len(b)\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = b[j-1] + bb\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            aa = a[i-1] + aa\n",
    "            bb = '-' + bb\n",
    "            i -= 1\n",
    "        else:\n",
    "            aa = '-' + aa\n",
    "            bb = b[j-1] + bb\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        aa = '-'*j + aa\n",
    "        bb = b[:j] + bb\n",
    "    elif j == 0:\n",
    "        aa = a[:i] + aa\n",
    "        bb = '-'*i + bb\n",
    "\n",
    "    return aa, bb\n",
    "\n",
    "def needleman_wunsch_global_alignment(a:str, b:str, sim_fct=sim_blossum, gap_cost=-8):\n",
    "    \"\"\" Global alignment with linear gap cost \"\"\"\n",
    "    M, D = compute_matrix_needleman_wunsch(a, b, sim_fct, gap_cost)\n",
    "    aa, bb = trace_back(a, b, D)\n",
    "    print(aa)\n",
    "    print(bb)\n",
    "    return M, aa, bb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gotoh: alignment with affine gap cost"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of one matrix, we need 3:\n",
    "- $F_{i,j}$ storing the best score for $a_{1...i}$ and $b_{1...j}$ that ends with $a_i$ and $b_j$\n",
    "- $A_{i,j}$ storing the best score that ends with $a_i$ and a gap\n",
    "- $B_{i,j}$ storing the best score that ends with a gap and $b_j$\n",
    "\n",
    "We have:\n",
    "$$\n",
    "A_{i,j} = \\max \\begin{cases}\n",
    "  A_{i-1,j} + \\text{gap extension} \\\\\n",
    "  F_{i-1,j} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "$$\n",
    "B_{i,j} = \\max \\begin{cases}\n",
    "  B_{i,j-1} + \\text{gap extension} \\\\\n",
    "  F_{i,j-1} + \\text{gap opening} &\\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "$$\n",
    "F_{i,j} = \\max \\begin{cases}\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "  F_{i-1,j-1} + s(a_i,b_j)  \\\\\n",
    "\\end{cases}\n",
    "$$\n",
    "\n",
    "The matrix $M_{i,j}$ as defined previously is the max (element wise) of this three matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_matrix_gotoh(seq_a:str, seq_b:str, sim_fct, gap_opening:float=-2, gap_extension:float=-2):\n",
    "    \"\"\" compute the Gotoh matrices \"\"\"\n",
    "\n",
    "    n_a = len(seq_a)\n",
    "    n_b = len(seq_b)\n",
    "    \n",
    "    F = np.zeros((n_a+1, n_b+1))\n",
    "    A = np.zeros((n_a+1, n_b+1))\n",
    "    B = np.zeros((n_a+1, n_b+1))\n",
    "\n",
    "    # init matrix A\n",
    "    for j in range(0, n_b+1):\n",
    "        A[0,j] = -np.inf\n",
    "        A[1,j] = 2*gap_opening + (j-1)*gap_extension\n",
    "    for i in range(1, n_a+1):    \n",
    "        A[i,0] = gap_opening + (i-1)*gap_extension\n",
    "\n",
    "    # init matrix B\n",
    "    for i in range(0, n_a+1):\n",
    "        B[i,0] = -np.inf\n",
    "        B[i,1] = 2*gap_opening + (i-1)*gap_extension\n",
    "    for j in range(1, n_b+1):\n",
    "        B[0,j] = gap_opening + (j-1)*gap_extension\n",
    "\n",
    "    # init matrix F\n",
    "    F[0,0] = 0\n",
    "    for i in range(1, n_a+1):\n",
    "        F[i,0] = -np.inf\n",
    "    for j in range(1, n_b+1):\n",
    "        F[0,j] = -np.inf\n",
    "        F[0,j] = -np.inf    \n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if i>1:\n",
    "                A[i,j] = max(F[i-1,j] + gap_opening, A[i-1,j] + gap_extension)\n",
    "            if j>1:\n",
    "                B[i,j] = max(F[i,j-1] + gap_opening, B[i,j-1] + gap_extension)\n",
    "            sim = sim_fct(seq_a[i-1], seq_b[j-1])\n",
    "            F[i,j] = max(F[i-1, j-1] + sim, A[i-1,j-1] + sim, B[i-1,j-1] + sim)\n",
    "\n",
    "    # get score and traceback matrix\n",
    "    M = np.maximum(F, np.maximum(A,B))\n",
    "    D = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if M[i,j] == F[i,j]:\n",
    "                D[i,j] = 1\n",
    "            elif M[i,j] == A[i,j]:\n",
    "                D[i,j] = 2\n",
    "            else:\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def gotoh_global_alignment(seq_a:str, seq_b:str, sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Global alignment with affine gap cost \"\"\"\n",
    "    M, D = compute_matrix_gotoh(seq_a, seq_b, sim_fct, gap_opening, gap_extension)\n",
    "    seq_aa, seq_bb = trace_back(seq_a, seq_b, D)\n",
    "    # print(seq_aa)\n",
    "    # print(seq_bb)\n",
    "    return M, seq_aa, seq_bb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[  0.  -3.  -4.  -5.  -6.]\n",
      " [ -3.   1.  -2.  -3.  -4.]\n",
      " [ -4.  -2.   0.  -3.  -4.]\n",
      " [ -5.  -3.  -1.  -1.  -4.]\n",
      " [ -6.  -4.  -4.  -2.  -2.]\n",
      " [ -7.  -5.  -3.  -5.  -3.]\n",
      " [ -8.  -6.  -6.  -4.  -6.]\n",
      " [ -9.  -7.  -7.  -5.  -5.]\n",
      " [-10.  -8.  -8.  -8.  -4.]]\n",
      "[[  0.  -3.  -6.  -9. -12.]\n",
      " [ -3.   1.  -2.  -5.  -8.]\n",
      " [ -6.  -2.   0.  -3.  -6.]\n",
      " [ -9.  -5.  -1.  -1.  -4.]\n",
      " [-12.  -8.  -4.  -2.  -2.]\n",
      " [-15. -11.  -7.  -5.  -3.]\n",
      " [-18. -14. -10.  -8.  -6.]\n",
      " [-21. -17. -13.  -9.  -9.]\n",
      " [-24. -20. -16. -12.  -8.]]\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "align_a = \"CHCHCHAT\"\n",
    "align_b = \"CCAT\"\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_basic, -3, -1)\n",
    "print(M)\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_basic, -3, -3)\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pairwise profile alignment\n",
    "We want to merge two alignments $A$ and $B$ into one by aligning them. To compute the similarity matrix, we just have to modify the previous algorithms by replacing the similarity cost of two bases by the average similarity over all the bases in the two alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def append(aa, a):\n",
    "    if type(aa) == str:     # aa, a = str, chr\n",
    "        aa += a\n",
    "    elif type(a) == list:   # aa, a = list of tuple, list of tuple\n",
    "        aa.extend(a)\n",
    "    else:\n",
    "        aa.append(a)\n",
    "    return aa\n",
    "\n",
    "def compute_matrix_gotoh_multiple(align_a:List[str], align_b:List[str], sim_fct=sim_basic, gap_opening:float=-2, gap_extension:float=-2):\n",
    "    \"\"\" compute the Gotoh matrices \"\"\"\n",
    "\n",
    "    n_a = len(align_a[0])\n",
    "    n_b = len(align_b[0])\n",
    "    \n",
    "    F = np.zeros((n_a+1, n_b+1))\n",
    "    A = np.zeros((n_a+1, n_b+1))\n",
    "    B = np.zeros((n_a+1, n_b+1))\n",
    "\n",
    "    # init matrix A\n",
    "    for j in range(0, n_b+1):\n",
    "        A[0,j] = -np.inf\n",
    "        A[1,j] = 2*gap_opening + (j-1)*gap_extension\n",
    "    for i in range(1, n_a+1):    \n",
    "        A[i,0] = gap_opening + (i-1)*gap_extension\n",
    "\n",
    "    # init matrix B\n",
    "    for i in range(0, n_a+1):\n",
    "        B[i,0] = -np.inf\n",
    "        B[i,1] = 2*gap_opening + (i-1)*gap_extension\n",
    "    for j in range(1, n_b+1):\n",
    "        B[0,j] = gap_opening + (j-1)*gap_extension\n",
    "\n",
    "    # init matrix F\n",
    "    F[0,0] = 0\n",
    "    for i in range(1, n_a+1):\n",
    "        F[i,0] = -np.inf\n",
    "    for j in range(1, n_b+1):\n",
    "        F[0,j] = -np.inf\n",
    "        F[0,j] = -np.inf    \n",
    "\n",
    "    # fill matrices\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if i>1:\n",
    "                A[i,j] = max(F[i-1,j] + gap_opening, A[i-1,j] + gap_extension)\n",
    "            if j>1:\n",
    "                B[i,j] = max(F[i,j-1] + gap_opening, B[i,j-1] + gap_extension)\n",
    "\n",
    "            # sim = average similarity between the two sets of aligned sequences\n",
    "            # problem: how do we handle gaps that might already exists in the alignment ?\n",
    "            sim = 0\n",
    "            for a in align_a:\n",
    "                for b in align_b:\n",
    "                    if a[i-1] == '-' and b[j-1] == '-':\n",
    "                            sim += 0\n",
    "                    elif a[i-1] == '-':\n",
    "                        if i>0 and a[i-1] == '-':\n",
    "                            sim += gap_extension\n",
    "                        else:\n",
    "                            sim += gap_opening            \n",
    "                    elif b[j-1] == '-':\n",
    "                        if j>0 and b[j-1] == '-':\n",
    "                            sim += gap_extension\n",
    "                        else:\n",
    "                            sim += gap_opening\n",
    "                    else:\n",
    "                        sim += sim_fct(a[i-1], b[j-1])\n",
    "            sim /= len(align_a) * len(align_b)\n",
    "\n",
    "            F[i,j] = max(F[i-1, j-1] + sim, A[i-1,j-1] + sim, B[i-1,j-1] + sim)\n",
    "\n",
    "    # get score and traceback matrix\n",
    "    M = np.maximum(F, np.maximum(A,B))\n",
    "    D = np.zeros((n_a+1, n_b+1), dtype=int)\n",
    "    for i in range(1, n_a+1):\n",
    "        for j in range(1, n_b+1):\n",
    "            if M[i,j] == F[i,j]:\n",
    "                D[i,j] = 1\n",
    "            elif M[i,j] == A[i,j]:\n",
    "                D[i,j] = 2\n",
    "            else:\n",
    "                D[i,j] = 3\n",
    "    \n",
    "    return M, D\n",
    "\n",
    "def trace_back_multiple(align_a:List[str], align_b:List[str], D:np.array):\n",
    "    \"\"\" Determine the alignment given a matrix with directions \"\"\"\n",
    "\n",
    "    if type(align_a[0]) == str:\n",
    "        align_aa = [\"\" for _ in range(len(align_a))]\n",
    "        align_bb = [\"\" for _ in range(len(align_b))]\n",
    "    else:\n",
    "        align_aa = [[] for _ in range(len(align_a))]\n",
    "        align_bb = [[] for _ in range(len(align_b))]\n",
    "    \n",
    "    i = len(align_a[0])\n",
    "    j = len(align_b[0])\n",
    "    \n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = append(align_aa[a], align_a[a][i-1])\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = append(align_bb[b], align_b[b][j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = append(align_aa[a], align_a[a][i-1])\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = append(align_bb[b], '-')\n",
    "            i -= 1\n",
    "        else:\n",
    "            for a in range(len(align_a)):\n",
    "                align_aa[a] = append(align_aa[a], '-')\n",
    "            for b in range(len(align_b)):\n",
    "                align_bb[b] = append(align_bb[b], align_b[b][j-1])\n",
    "            j -= 1\n",
    "    while j>0:\n",
    "        for a in range(len(align_a)):\n",
    "            align_aa[a] = append(align_aa[a],  '-')\n",
    "        for b in range(len(align_b)):\n",
    "            align_bb[b] = append(align_bb[b], align_b[b][j-1])\n",
    "            j -= 1\n",
    "    while i > 0:\n",
    "        for a in range(len(align_a)):\n",
    "            align_aa[a] = append(align_aa[a], align_a[a][i-1])\n",
    "        for b in range(len(align_b)):\n",
    "            align_bb[b] = append(align_bb[b], '-')\n",
    "        i -= 1\n",
    "\n",
    "    align = align_aa + align_bb    # list concatenation\n",
    "\n",
    "    return [seq[::-1] for seq in align]\n",
    " \n",
    "\n",
    "def gotoh_global_alignment_multiple(align_a:List[str], align_b:List[str], sim_fct=sim_basic, gap_opening=-2, gap_extension=-2):\n",
    "    \"\"\" Global alignment with affine gap cost \"\"\"\n",
    "    M, D = compute_matrix_gotoh_multiple(align_a, align_b, sim_fct, gap_opening, gap_extension)\n",
    "    align = trace_back_multiple(align_a, align_b, D)\n",
    "    return M, align"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0. -2. -3. -4.]\n",
      " [-2.  1. -1. -2.]\n",
      " [-3. -1.  0. -2.]\n",
      " [-4. -2.  0. -1.]\n",
      " [-5. -3. -2.  1.]]\n",
      "[[ 0. -2. -3. -4.]\n",
      " [-2.  1. -1. -2.]\n",
      " [-3. -1.  0. -2.]\n",
      " [-4. -2.  0. -1.]\n",
      " [-5. -3. -2.  1.]]\n",
      "['CHAT', 'C-AT']\n",
      "['CHAT', 'C-AT', 'C-IT']\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "align_a = \"CHAT\"\n",
    "align_b = \"CAT\"\n",
    "seq_c = \"CIT\"\n",
    "\n",
    "M, aa, bb = gotoh_global_alignment(align_a, align_b, sim_basic, -2, -1)\n",
    "print(M)\n",
    "\n",
    "M, align = gotoh_global_alignment_multiple([align_a], [align_b], sim_basic, -2, -1)\n",
    "print(M)\n",
    "print(align)\n",
    "\n",
    "M, align = gotoh_global_alignment_multiple(align, [seq_c], sim_basic, -2, -1)\n",
    "print(align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('A', 1), ('B', 2), ('C', 3)]\n"
     ]
    }
   ],
   "source": [
    "test = [('A',1), ('B',2)]\n",
    "test2 = append(test, ('C',3))\n",
    "print(test2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Progressive multiple alignment\n",
    "To align N sequences, we do clustering. First, we create one cluster per sequence. Then, we align and merge the two closest clusters until we only have one big cluster, containing all the alignment sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def progressive_alignment(seq:list, sim_fct=sim_blossum, gap_opening=-10, gap_extension=-0.5):\n",
    "  \"\"\" Computes multiple sequence alignement \"\"\"\n",
    "\n",
    "  # transforming each sequence into a cluster of one sequence\n",
    "  clusters = [[s] for s in seq]\n",
    "  # order of the clusters, to keep track of the original order\n",
    "  order = list(range(len(seq)))\n",
    "\n",
    "  # initialising distance matrix\n",
    "  matrix = np.zeros((len(seq), len(seq)))\n",
    "  for i in range(len(seq)):\n",
    "    for j in range(i+1, len(seq)):\n",
    "      matrix[i,j] = compute_matrix_gotoh(seq[i], seq[j], sim_fct, gap_opening, gap_extension)[0][-1,-1]\n",
    "      matrix[j,i] = matrix[i,j]\n",
    "\n",
    "  while len(clusters) > 1:\n",
    "    # finding closest clusters (cad with highest similarity)\n",
    "    max_sim = -np.inf\n",
    "    for i in range(len(clusters)):\n",
    "      for j in range(i+1,len(clusters)):\n",
    "        if matrix[i,j] > max_sim:\n",
    "          max_sim = matrix[i,j]\n",
    "          best_i = i\n",
    "          best_j = j\n",
    "\n",
    "    # merging two clusters best_i < best_j\n",
    "    cluster_j = clusters.pop(best_j)\n",
    "    cluster_i = clusters.pop(best_i)\n",
    "    \n",
    "    order_j = order.pop(best_j)\n",
    "    order_i = order.pop(best_i)\n",
    "    order.append([order_i, order_j])\n",
    "\n",
    "    M, D = compute_matrix_gotoh_multiple(cluster_i, cluster_j, sim_fct, gap_opening, gap_extension)\n",
    "    new_cluster = trace_back_multiple(cluster_i, cluster_j, D)\n",
    "\n",
    "    clusters.append(new_cluster)\n",
    "\n",
    "    # updating distance matrix\n",
    "    new_line = (len(cluster_i)*matrix[best_i,:] + len(cluster_j)*matrix[best_j,:]) / (len(cluster_i)+len(cluster_j))  # weighted average\n",
    "    new_column = np.append(new_line, 0)  # adding a column of 0\n",
    "    new_line = new_line.reshape(1, -1)\n",
    "    new_column = new_column.reshape(-1, 1)\n",
    "\n",
    "    matrix = np.append(matrix, new_line, axis=0)\n",
    "    matrix = np.append(matrix, new_column, axis=1)\n",
    "\n",
    "    matrix = np.delete(matrix, [best_i, best_j], axis=0)\n",
    "    matrix = np.delete(matrix, [best_i, best_j], axis=1)\n",
    "\n",
    "  def get_order(gt):\n",
    "    \"\"\"((4, (0, 1)), (2, 3)) -> [4, 0, 1, 2, 3]\"\"\"\n",
    "    if type(gt) == int:\n",
    "        return [gt]\n",
    "    else:\n",
    "        return get_order(gt[0]) + get_order(gt[1])\n",
    "    \n",
    "  order = get_order(order[0])\n",
    "  reverse_order = [order.index(i) for i in range(len(order))]\n",
    "  \n",
    "  alignment = clusters[0]\n",
    "  alignment = [alignment[i] for i in reverse_order]\n",
    "  \n",
    "  return alignment, M[-1,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-11.5\n",
      "['-CCCAT', 'THHHAT', '---HER']\n"
     ]
    }
   ],
   "source": [
    "# test: easy example\n",
    "seq = [\"CCCAT\", \"CHCHCHAT\", \"HER\"]\n",
    "\n",
    "align, score = progressive_alignment(seq, sim_blossum)\n",
    "print(score)\n",
    "print(align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MALSAEDRALVRALWKKLGSNVGVYTTEALERTFLAFPATKTYFSHLDLSPGSSQVRAHGQKVADALSLAVERLDDLPHALSALSHLHACQLRVDPASFQLLGHCLLVTLARHYPGDFSPALQASLDKFLSHVISALVSEYR\n",
      "MSLTKTERTIIVSMWAKISTQADTIGTETLERLFLSHPQTKTYFPHFDLHPGSAQLRAHGSKVVAAVGDAVKSIDDIGGALSKLSELHAYILRVDPVNFKLLSHCLLVTLAARFPADFTAEAHAAWDKFLSVVSSVLTEKYR\n",
      "MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHFDLSHGSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKYR\n",
      "MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKYH\n",
      "MVHLTPEEKTAVNALWGKVNVDAVGGEALGRLLVVYPWTQRFFESFGDLSSPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFSQLSELHCDKLHVDPENFRLLGNVLVCVLARNFGKEFTPQMQAAYQKVVAGVANALAHKYH\n",
      "MVHFTAEEKAAVTSLWSKMNVEEAGGEALGRLLVVYPWTQRFFDSFGNLSSPSAILGNPKVKAHGKKVLTSFGDAIKNMDNLKPAFAKLSELHCDKLHVDPENFKLLGNVMVIILATHFGKEFTPEVQAAWQKLVSAVAIALAHKYH\n",
      "MGHFTEEDKATITSLWGKVNVEDAGGETLGRLLVVYPWTQRFFDSFGNLSSASAIMGNPKVKAHGKKVLTSLGDAIKHLDDLKGTFAQLSELHCDKLHVDPENFKLLGNVLVTVLAIHFGKEFTPEVQASWQKMVTAVASALSSRYH\n",
      "MGLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "MA-LSAEDRALVRALWKKLGSNVGVYTTEALERTFLAFPATKTYFSHL-DLS--P---GSSQVRAHGQKVADALSLAVERLDDLPHALSALSHLHACQLRVDPASFQLLGHCLLVTLARHYPGDFSPALQASLDKFLSHVISALVSEY-----R-\n",
      "MS-LTKTERTIIVSMWAKISTQADTIGTETLERLFLSHPQTKTYFPHF-DLH--P---GSAQLRAHGSKVVAAVGDAVKSIDDIGGALSKLSELHAYILRVDPVNFKLLSHCLLVTLAARFPADFTAEAHAAWDKFLSVVSSVLTEKY-----R-\n",
      "MV-LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHF-DLS--H---GSAQVKGHGKKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTPAVHASLDKFLASVSTVLTSKY-----R-\n",
      "MVHLTPEEKSAVTALWGKV--NVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFGKEFTPPVQAAYQKVVAGVANALAHKY-----H-\n",
      "MVHLTPEEKTAVNALWGKV--NVDAVGGEALGRLLVVYPWTQRFFESFGDLSSPDAVMGNPKVKAHGKKVLGAFSDGLAHLDNLKGTFSQLSELHCDKLHVDPENFRLLGNVLVCVLARNFGKEFTPQMQAAYQKVVAGVANALAHKY-----H-\n",
      "MVHFTAEEKAAVTSLWSKM--NVEEAGGEALGRLLVVYPWTQRFFDSFGNLSSPSAILGNPKVKAHGKKVLTSFGDAIKNMDNLKPAFAKLSELHCDKLHVDPENFKLLGNVMVIILATHFGKEFTPEVQAAWQKLVSAVAIALAHKY-----H-\n",
      "MGHFTEEDKATITSLWGKV--NVEDAGGETLGRLLVVYPWTQRFFDSFGNLSSASAIMGNPKVKAHGKKVLTSLGDAIKHLDDLKGTFAQLSELHCDKLHVDPENFKLLGNVLVTVLAIHFGKEFTPEVQASWQKMVTAVASALSSRY-----H-\n",
      "MG-LSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "Score of alignment: 119.07142857142856\n",
      "Size of alignment: 155\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# test: real example\n",
    "seq, ids = read_fasta(\"data/paralogs.fasta\")\n",
    "# seq = read_fasta(\"data/paralogs.fasta\")\n",
    "\n",
    "for s in seq:\n",
    "  print(s)\n",
    "print()\n",
    "\n",
    "alignment, score = progressive_alignment(seq, sim_blossum, gap_opening=-10, gap_extension=-0.5)\n",
    "for s in alignment:\n",
    "  print(s)\n",
    "print(f\"Score of alignment: {score}\")\n",
    "print(f\"Size of alignment: {len(alignment[0])}\")\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Include structural information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and read PDB files\n",
    "`DSSP(model, PDBfile)` is an algorithm to *Define Secondary Structure of Proteins*. It returns a dictionary describing the structure of the protein. In this dictionary:\n",
    "- the keys are tupple (chain_id, residue_id)\n",
    "- the values are tupple (DSSP index, Amino acid, Secondary structure, solvent accessibility, Phi, Psi, NH–>O_1_relidx,...) (see https://biopython.org/docs/1.75/api/Bio.PDB.DSSP.html)\n",
    "    - for example: (3, 'V', '-', 0.7394366197183099, -62.9, 147.7, 1, -0.1, 2, -1.6, 0, 0.0, -1, -0.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ssp(mol:str):\n",
    "    # download PDB file\n",
    "    PDBfile = f\"data/pdb/{mol}.pdb\"\n",
    "    if not os.path.exists(PDBfile):\n",
    "        pdbl = PDBList()\n",
    "        oldPDBfile = pdbl.retrieve_pdb_file(mol, pdir='data/pdb/', file_format=\"pdb\")\n",
    "        os.rename(oldPDBfile, PDBfile)\n",
    "\n",
    "    # parse PDB file\n",
    "    parser = PDBParser()\n",
    "    with warnings.catch_warnings():\n",
    "        structure = parser.get_structure(mol, PDBfile)\n",
    "\n",
    "    # get DSSP\n",
    "    model = structure[0]\n",
    "    ssp = DSSP(model, PDBfile)\n",
    "    return model, ssp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chain A\n",
      "(1, 'M', '-', 1.0, 360.0, -66.0, 0, 0.0, 0, 0.0, 0, 0.0, 0, 0.0)\n",
      "(2, 'E', '-', 0.7628865979381443, 64.4, 100.8, 2, -0.1, 3, -0.0, 1, -0.0, 0, 0.0)\n",
      "(3, 'V', '-', 0.7394366197183099, -62.9, 147.7, 1, -0.1, 2, -1.6, 0, 0.0, -1, -0.0)\n",
      "(4, 'K', '-', 0.9414634146341463, -65.1, 89.4, 2, -0.0, 2, -0.2, 3, -0.0, -1, -0.1)\n",
      "(5, 'K', '-', 0.4926829268292683, -89.1, 156.7, -2, -1.6, 2, -1.9, 1, -0.1, -1, -0.0)\n",
      "(6, 'T', '-', 0.6197183098591549, -70.5, 84.3, -2, -0.2, 2, -1.3, 1, -0.2, -1, -0.1)\n"
     ]
    }
   ],
   "source": [
    "# test        \n",
    "mol = \"1idy\"\n",
    "model, ssp = get_ssp(mol)\n",
    "\n",
    "for chain in model:\n",
    "    print(f\"chain {chain.id}\")\n",
    "    i = 0\n",
    "    for residue in chain:\n",
    "        print(ssp[(chain.id, residue.id)])\n",
    "        i += 1\n",
    "        if i > 5:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta_and_PDB(filename) -> Tuple[List[str], List[List[Tuple]]]:\n",
    "    \"\"\"\n",
    "    Read a fasta file and the corresponding PDB files\n",
    "    Returns:\n",
    "    - seqFASTA, list of str: the sequences in the fasta file\n",
    "    - SSPs, list of list of tuple: list of the secondary structure infos for the corresponding molecules\n",
    "    \"\"\"\n",
    "    seqFASTA = []\n",
    "    ids = []\n",
    "    SSPs = []\n",
    "    for record in SeqIO.parse(filename, \"fasta\"):\n",
    "\n",
    "        seqFASTA.append(str(record.seq))\n",
    "        ids.append(record.id)\n",
    "\n",
    "        name = record.name\n",
    "        mol, chain = name.split(\"_\")\n",
    "        if chain == \"\":\n",
    "            chain = \"A\"\n",
    "        model, ssp = get_ssp(mol)\n",
    "\n",
    "        # transforms ssp dict into a list containing only the amino acids\n",
    "        new_ssp = []\n",
    "        for residue in model[chain]:\n",
    "            if is_aa(residue):\n",
    "                try:\n",
    "                    new_ssp.append(ssp[(chain, residue.id)])                    \n",
    "                except KeyError:\n",
    "                    print(f\"residue {residue} not found in DSSP for {mol}_{chain}\")\n",
    "        SSPs.append(new_ssp)\n",
    "    return seqFASTA, SSPs, ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align fasta and PDB\n",
    "Some amino-acids might be missing in the PDB file. Therefore we have to use Needleman-Wunsch to align properly the sequence obtained from the fasta and from the PDB file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trace_back_SSP(seq:str, ssp:List[tuple], D:np.array) -> Tuple[str, List[tuple]]:\n",
    "    \"\"\" traceback matrix D to align a sequence and a list of ssp lines \"\"\"\n",
    "    new_seq = \"\"\n",
    "    new_ssp = []\n",
    "    \n",
    "    i = len(seq)\n",
    "    j = len(ssp)\n",
    "    while i>0 and j>0 :\n",
    "        if D[i,j] == 1:\n",
    "            new_seq += seq[i-1]\n",
    "            new_ssp.append(ssp[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif D[i,j] == 2:   # a base is missing in the PDB\n",
    "            new_seq += seq[i-1]\n",
    "            missing_res = ['-' for _ in range(14)] # we don't know the secondary structure\n",
    "            missing_res[1] = seq[i-1]               # but we know the amino acid\n",
    "            new_ssp.append(tuple(missing_res)) \n",
    "            i -= 1\n",
    "        else:               # a base is missing in the sequence but present in the PDB (not expected)\n",
    "            new_seq += ssp[j-1][1]\n",
    "            new_ssp.append(ssp[j-1])\n",
    "            j -= 1\n",
    "    if i == 0:\n",
    "        new_seq += '-'*j\n",
    "        new_ssp += ssp[:j]\n",
    "    elif j == 0:\n",
    "        new_seq += seq[:i]\n",
    "        new_ssp += [tuple() for _ in range(i)]     \n",
    "    return new_seq[::-1], new_ssp[::-1]\n",
    "\n",
    "def align_SSP(seq:str, dssp:List[tuple]):\n",
    "    \"\"\" align sequence and list of PDB lines\"\"\"\n",
    "    sim_bool = lambda a, b: 1 if a == b else 0\n",
    "    M, D = compute_matrix_needleman_wunsch(seq, \"\".join([d[1] for d in dssp]), sim_bool, gap_cost=-1)\n",
    "    new_seq, new_dssp = trace_back_SSP(seq, dssp, D)\n",
    "    return new_seq, new_dssp\n",
    "\n",
    "def read_and_align_fasta_and_SSP(filename):\n",
    "    seq_list, ssp_list, ids = read_fasta_and_PDB(filename)\n",
    "    new_seq_list, new_ssp_list = [], []\n",
    "    for i in range(len(seq_list)):\n",
    "        seq, ssp = align_SSP(seq_list[i], ssp_list[i])\n",
    "        new_seq_list.append(seq)\n",
    "        new_ssp_list.append(ssp)\n",
    "    return new_seq_list, new_ssp_list, ids\n",
    "\n",
    "def ssp2tupple(ssp:List[tuple]) -> List[Tuple[str, float]]:\n",
    "    \"\"\" extracts a list of tuples (amino-acid, burial) from a list of ssp lines \"\"\"\n",
    "    return [(res[1], res[3]) for res in ssp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9025.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9069.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9113.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9240.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8685.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8701.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8733.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9485.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15917.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15948.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15979.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16010.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 16041.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 16142.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 16238.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16349.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue MSE het=H_MSE resseq=78 icode= > not found in DSSP for 1o20_A\n",
      "residue <Residue MSE het=H_MSE resseq=189 icode= > not found in DSSP for 1o20_A\n"
     ]
    }
   ],
   "source": [
    "# test\n",
    "seq_list, ssp_list, ids = read_and_align_fasta_and_SSP(\"data/balibase/RV11.unaligned/BB11003.fasta\")\n",
    "burial_list = [ssp2tupple(ssp) for ssp in ssp_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SISDTVKRAREAFNSGKTRSLQFRIQQLEALQRMINENLKSISGALASDLGKNEWTSYYEEVAHVLEELDTTIKELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLTIQPMVGAVAAGNAVILKPSEVSGHMADLLATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTPVTLELGGKSPCYVDKDCDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRDYGRIINDRHFQRVKGLIDNQKVAHGGTWDQSSRYIAPTILVDVDPQSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNEKVIKKMIAETSSGGVTANDVIVHITVPTLPFGGVGNSGMGAYHGKKSFETFSHRRSCLVKSLLNEEAHKARYPPSPA\n",
      "SISDTVKRAREAFNSGKTRSLQFRIQQLEALQRMINENLKSISGALASDLGKNEWTSYYEEVAHVLEELDTTIKELPDWAEDEPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLTIQPMVGAVAAGNAVILKPSEVSGHMADLLATLIPQYMDQNLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTPVTLELGGKSPCYVDKDCDLDVACRRIAWGKFMNSGQTCVAPDYILCDPSIQNQIVEKLKKSLKDFYGEDAKQSRDYGRIINDRHFQRVKGLIDNQKVAHGGTWDQSSRYIAPTILVDVDPQSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALYVFSNNEKVIKKMIAETSSGGVTANDVIVHITVPTLPFGGVGNSGMGAYHGKKSFETFSHRRSCLVKSLLNEEAHKARYPPSPA\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEATLVYEVGKNWVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGEDNESFYVPLGAGVVIAPWNFPVAIFTGMIVGPVAVGNTVIAKPAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYLVEHPRIRFINFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIVDETADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPDLGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVFTEVPPKARIAQEEIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKREHLEWARREFHVGNLYFNRKITGALVGVQPFGGFKLSGTNAKTGALDYLRLFLEMKAVAERF\n",
      "MTVEPFRNEPIETFQTEEARRAMREALRRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEAALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEATLVYEVGKNWVEASADVAEAIDFIEYYARAALRYRYPAVEVVPYPGEDNESFYVPLGAGVVIAPWNFPVAIFTGMIVGPVAVGNTVIAKPAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYLVEHPRIRFINFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETGGKNAIIVDETADFDLAAEGVVVSAYGFQGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGPAEENPDLGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVFTEVPPKARIAQEEIFGPVLSVIRVKDFAEALEVANDTPYGLTGGVYSRKREHLEWARREFHVGNLYFNRKITGALVGVQPFGGFKLSGTNAKTGALDYLRLFLEMKAVAERF\n",
      "TDNVFYATNAFTGEALPLAFPVHTEVEVNQAATAAAKVARDFRRLNNSKRASLLRTIASELEARSDDIIARAHLETALPEVRLTGEIARTANQLRLFADVVNSGSYHQAILDTPNPTRAPLPKPDIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGTSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQALVSHPEIKAVGFTGSVGGGRALFNLAHERPEPIPFYGELGAINPTFIFPSAMRAKADLADQFVASMTMGCGQFCTKPGVVFALNTPETQAFIETAQSLIRQQSPSTLLTPGIRDSYQSQVVSRGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "TDNVFYATNAFTGEALPLAFPVHTEVEVNQAATAAAKVARDFRRLNNSKRASLLRTIASELEARSDDIIARAHLETALPEVRLTGEIARTANQLRLFADVVNSGSYHQAILDTPNPTRAPLPKPDIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPGTSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQALVSHPEIKAVGFTGSVGGGRALFNLAHERPEPIPFYGELGAINPTFIFPSAMRAKADLADQFVASMTMGCGQFCTKPGVVFALNTPETQAFIETAQSLIRQQSPSTLLTPGIRDSYQSQVVSRGSDDGIDVTFSQAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEMLAGSLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "DELLEKAKKVREAWDVLRNATTREKNKAIKKIAEKLDERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDEXIKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGIIYESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSAIREALKETEIPESSVEFIENTDRSLVLEXIRLREYLSLVIPRGGYGLISFVRDNATVPVLETGVGNCHIFVDESADLKKAVPVIINAKTQRPGTCNAAEKLLVHEKIAKEFLPVIVEELRKHGVEVRGCEKTREIVPDVVPATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESILTENYSNAKKFVSEIDAAAVYVNASTRFTDGGQFGFGAEIGISTQRFHARGPVGLRELTTYKFVVLGEYHVRE\n",
      "DELLEKAKKVREAWDVLRNATTREKNKAIKKIAEKLDERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDEXIKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGIIYESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSAIREALKETEIPESSVEFIENTDRSLVLEXIRLREYLSLVIPRGGYGLISFVRDNATVPVLETGVGNCHIFVDESADLKKAVPVIINAKTQRPGTCNAAEKLLVHEKIAKEFLPVIVEELRKHGVEVRGCEKTREIVPDVVPATEDDWPTEYLDLIIAIKVVKNVDEAIEHIKKYSTGHSESILTENYSNAKKFVSEIDAAAVYVNASTRFTDGGQFGFGAEIGISTQRFHARGPVGLRELTTYKFVVLGEYHVRE\n",
      "\n",
      "[('S', 0.8461538461538461), ('I', 0.10650887573964497), ('S', 0.24615384615384617), ('D', 0.5030674846625767), ('T', 0.14084507042253522), ('V', 0.04225352112676056), ('K', 0.5365853658536586), ('R', 0.41935483870967744), ('A', 0.0), ('R', 0.24193548387096775), ('E', 0.711340206185567), ('A', 0.11320754716981132), ('F', 0.1065989847715736), ('N', 0.7070063694267515), ('S', 0.5538461538461539), ('G', 0.34523809523809523), ('K', 0.3804878048780488), ('T', 0.0), ('R', 0.375), ('S', 0.33076923076923076), ('L', 0.24390243902439024), ('Q', 0.6515151515151515), ('F', 0.14213197969543148), ('R', 0.06451612903225806), ('I', 0.20118343195266272), ('Q', 0.5656565656565656), ('Q', 0.10606060606060606), ('L', 0.0), ('E', 0.3711340206185567), ('A', 0.018867924528301886), ('L', 0.0), ('Q', 0.23737373737373738), ('R', 0.4596774193548387), ('M', 0.0), ('I', 0.0), ('N', 0.5605095541401274), ('E', 0.6443298969072165), ('N', 0.17197452229299362), ('L', 0.22560975609756098), ('K', 0.6292682926829268), ('S', 0.4076923076923077), ('I', 0.0), ('S', 0.07692307692307693), ('G', 0.40476190476190477), ('A', 0.09433962264150944), ('L', 0.0), ('A', 0.3018867924528302), ('S', 0.47692307692307695), ('D', 0.049079754601226995), ('L', 0.0), ('G', 0.08333333333333333), ('K', 0.03414634146341464), ('N', 0.11464968152866242), ('E', 0.5824742268041238), ('W', 0.15418502202643172), ('T', 0.04929577464788732), ('S', 0.0), ('Y', 0.3738738738738739), ('Y', 0.5990990990990991), ('E', 0.13402061855670103), ('E', 0.0), ('V', 0.007042253521126761), ('A', 0.20754716981132076), ('H', 0.27717391304347827), ('V', 0.0), ('L', 0.13414634146341464), ('E', 0.3711340206185567), ('E', 0.020618556701030927), ('L', 0.0), ('D', 0.4662576687116564), ('T', 0.21830985915492956), ('T', 0.014084507042253521), ('I', 0.13609467455621302), ('K', 0.6780487804878049), ('E', 0.4690721649484536), ('L', 0.0), ('P', 0.4411764705882353), ('D', 0.7607361963190185), ('W', 0.08370044052863436), ('A', 0.2358490566037736), ('E', 0.7164948453608248), ('D', 0.39263803680981596), ('E', 0.36597938144329895), ('P', 0.7794117647058824), ('V', 0.2112676056338028), ('A', 0.6037735849056604), ('K', 0.4975609756097561), ('T', 0.18309859154929578), ('R', 0.9274193548387096), ('Q', 0.5303030303030303), ('T', 0.02112676056338028), ('Q', 0.48484848484848486), ('Q', 0.7676767676767676), ('D', 0.07975460122699386), ('D', 0.49693251533742333), ('L', 0.042682926829268296), ('Y', 0.07207207207207207), ('I', 0.07100591715976332), ('H', 0.13043478260869565), ('S', 0.1), ('E', 0.02577319587628866), ('P', 0.03676470588235294), ('L', 0.08536585365853659), ('G', 0.0), ('V', 0.0), ('V', 0.0), ('L', 0.0), ('V', 0.007042253521126761), ('I', 0.0650887573964497), ('G', 0.09523809523809523), ('A', 0.14150943396226415), ('W', 0.14537444933920704), ('N', 0.03821656050955414), ('Y', 0.09009009009009009), ('P', 0.03676470588235294), ('F', 0.005076142131979695), ('N', 0.11464968152866242), ('L', 0.036585365853658534), ('T', 0.007042253521126761), ('I', 0.0), ('Q', 0.07575757575757576), ('P', 0.0), ('M', 0.0), ('V', 0.0), ('G', 0.03571428571428571), ('A', 0.0), ('V', 0.0), ('A', 0.0), ('A', 0.018867924528301886), ('G', 0.0), ('N', 0.0), ('A', 0.0), ('V', 0.0), ('I', 0.0), ('L', 0.012195121951219513), ('K', 0.02926829268292683), ('P', 0.03676470588235294), ('S', 0.09230769230769231), ('E', 0.42783505154639173), ('V', 0.3873239436619718), ('S', 0.007692307692307693), ('G', 0.30952380952380953), ('H', 0.5543478260869565), ('M', 0.005319148936170213), ('A', 0.07547169811320754), ('D', 0.5276073619631901), ('L', 0.1524390243902439), ('L', 0.012195121951219513), ('A', 0.24528301886792453), ('T', 0.5492957746478874), ('L', 0.024390243902439025), ('I', 0.0), ('P', 0.375), ('Q', 0.6666666666666666), ('Y', 0.1981981981981982), ('M', 0.03723404255319149), ('D', 0.19631901840490798), ('Q', 0.6363636363636364), ('N', 0.445859872611465), ('L', 0.0), ('Y', 0.009009009009009009), ('L', 0.10365853658536585), ('V', 0.2323943661971831), ('V', 0.09154929577464789), ('K', 0.44390243902439025), ('G', 0.2261904761904762), ('G', 0.16666666666666666), ('V', 0.6971830985915493), ('P', 0.6544117647058824), ('E', 0.37628865979381443), ('T', 0.16901408450704225), ('T', 0.528169014084507), ('E', 0.4639175257731959), ('L', 0.006097560975609756), ('L', 0.10975609756097561), ('K', 0.7024390243902439), ('E', 0.2422680412371134), ('R', 0.34274193548387094), ('F', 0.02030456852791878), ('D', 0.12883435582822086), ('H', 0.05434782608695652), ('I', 0.023668639053254437), ('M', 0.0), ('Y', 0.07207207207207207), ('T', 0.07746478873239436), ('G', 0.35714285714285715), ('S', 0.4076923076923077), ('T', 0.22535211267605634), ('A', 0.6509433962264151), ('V', 0.4225352112676056), ('G', 0.0), ('K', 0.2975609756097561), ('I', 0.6272189349112426), ('V', 0.02112676056338028), ('M', 0.010638297872340425), ('A', 0.330188679245283), ('A', 0.1792452830188679), ('A', 0.009433962264150943), ('A', 0.009433962264150943), ('K', 0.7073170731707317), ('H', 0.25), ('L', 0.012195121951219513), ('T', 0.04225352112676056), ('P', 0.25), ('V', 0.1267605633802817), ('T', 0.014084507042253521), ('L', 0.0), ('E', 0.015463917525773196), ('L', 0.024390243902439025), ('G', 0.08333333333333333), ('G', 0.27380952380952384), ('K', 0.13658536585365855), ('S', 0.023076923076923078), ('P', 0.014705882352941176), ('C', 0.0), ('Y', 0.03153153153153153), ('V', 0.0), ('D', 0.12883435582822086), ('K', 0.34146341463414637), ('D', 0.5398773006134969), ('C', 0.08148148148148149), ('D', 0.43558282208588955), ('L', 0.08536585365853659), ('D', 0.48466257668711654), ('V', 0.14084507042253522), ('A', 0.0), ('C', 0.0), ('R', 0.15725806451612903), ('R', 0.008064516129032258), ('I', 0.0), ('A', 0.0), ('W', 0.13215859030837004), ('G', 0.0), ('K', 0.004878048780487805), ('F', 0.005076142131979695), ('M', 0.09042553191489362), ('N', 0.012738853503184714), ('S', 0.007692307692307693), ('G', 0.011904761904761904), ('Q', 0.0), ('T', 0.056338028169014086), ('C', 0.18518518518518517), ('V', 0.04929577464788732), ('A', 0.07547169811320754), ('P', 0.0), ('D', 0.03680981595092025), ('Y', 0.018018018018018018), ('I', 0.0), ('L', 0.0), ('C', 0.0), ('D', 0.25766871165644173), ('P', 0.3382352941176471), ('S', 0.5307692307692308), ('I', 0.05917159763313609), ('Q', 0.015151515151515152), ('N', 0.6496815286624203), ('Q', 0.4595959595959596), ('I', 0.0), ('V', 0.035211267605633804), ('E', 0.5412371134020618), ('K', 0.2097560975609756), ('L', 0.006097560975609756), ('K', 0.2), ('K', 0.45365853658536587), ('S', 0.06153846153846154), ('L', 0.042682926829268296), ('K', 0.424390243902439), ('D', 0.5521472392638037), ('F', 0.015228426395939087), ('Y', 0.05855855855855856), ('G', 0.4642857142857143), ('E', 0.9020618556701031), ('D', 0.4233128834355828), ('A', 0.08490566037735849), ('K', 0.4292682926829268), ('Q', 0.7222222222222222), ('S', 0.18461538461538463), ('R', 0.5604838709677419), ('D', 0.06134969325153374), ('Y', 0.036036036036036036), ('G', 0.11904761904761904), ('R', 0.25), ('I', 0.0), ('I', 0.047337278106508875), ('N', 0.2802547770700637), ('D', 0.6380368098159509), ('R', 0.6693548387096774), ('H', 0.14130434782608695), ('F', 0.005076142131979695), ('Q', 0.43434343434343436), ('R', 0.2217741935483871), ('V', 0.0), ('K', 0.17073170731707318), ('G', 0.38095238095238093), ('L', 0.042682926829268296), ('I', 0.07692307692307693), ('D', 0.6012269938650306), ('N', 1.0), ('Q', 0.15151515151515152), ('K', 0.5756097560975609), ('V', 0.47183098591549294), ('A', 0.39622641509433965), ('H', 0.30978260869565216), ('G', 0.36904761904761907), ('G', 0.32142857142857145), ('T', 0.6619718309859155), ('W', 0.44933920704845814), ('D', 0.2883435582822086), ('Q', 0.5606060606060606), ('S', 0.8384615384615385), ('S', 0.36923076923076925), ('R', 0.23790322580645162), ('Y', 0.06756756756756757), ('I', 0.005917159763313609), ('A', 0.03773584905660377), ('P', 0.058823529411764705), ('T', 0.0), ('I', 0.0), ('L', 0.0), ('V', 0.14084507042253522), ('D', 0.558282208588957), ('V', 0.056338028169014086), ('D', 0.5644171779141104), ('P', 0.25), ('Q', 0.6717171717171717), ('S', 0.12307692307692308), ('P', 0.5514705882352942), ('V', 0.0), ('M', 0.0), ('Q', 0.51010101010101), ('E', 0.3711340206185567), ('E', 0.32989690721649484), ('I', 0.023668639053254437), ('F', 0.1319796954314721), ('G', 0.0), ('P', 0.0), ('V', 0.0), ('M', 0.0), ('P', 0.0), ('I', 0.0), ('V', 0.035211267605633804), ('C', 0.13333333333333333), ('V', 0.04225352112676056), ('R', 0.592741935483871), ('S', 0.3769230769230769), ('L', 0.17073170731707318), ('E', 0.5), ('E', 0.32989690721649484), ('A', 0.0), ('I', 0.047337278106508875), ('Q', 0.5656565656565656), ('F', 0.17766497461928935), ('I', 0.04142011834319527), ('N', 0.4585987261146497), ('Q', 0.7272727272727273), ('R', 0.28225806451612906), ('E', 0.4896907216494845), ('K', 0.21951219512195122), ('P', 0.03676470588235294), ('L', 0.012195121951219513), ('A', 0.0), ('L', 0.0), ('Y', 0.0), ('V', 0.0), ('F', 0.0), ('S', 0.0), ('N', 0.4267515923566879), ('N', 0.28662420382165604), ('E', 0.38144329896907214), ('K', 0.7268292682926829), ('V', 0.11971830985915492), ('I', 0.0), ('K', 0.3170731707317073), ('K', 0.424390243902439), ('M', 0.0), ('I', 0.01775147928994083), ('A', 0.46226415094339623), ('E', 0.41237113402061853), ('T', 0.014084507042253521), ('S', 0.09230769230769231), ('S', 0.0), ('G', 0.023809523809523808), ('G', 0.07142857142857142), ('V', 0.0), ('T', 0.0), ('A', 0.0), ('N', 0.0), ('D', 0.0), ('V', 0.0), ('I', 0.09467455621301775), ('V', 0.1267605633802817), ('H', 0.03260869565217391), ('I', 0.14792899408284024), ('T', 0.16901408450704225), ('V', 0.007042253521126761), ('P', 0.15441176470588236), ('T', 0.31690140845070425), ('L', 0.006097560975609756), ('P', 0.19852941176470587), ('F', 0.07614213197969544), ('G', 0.11904761904761904), ('G', 0.07142857142857142), ('V', 0.04929577464788732), ('G', 0.27380952380952384), ('N', 0.1464968152866242), ('S', 0.0), ('G', 0.03571428571428571), ('M', 0.13829787234042554), ('G', 0.47619047619047616), ('A', 0.09433962264150944), ('Y', 0.0), ('H', 0.043478260869565216), ('G', 0.023809523809523808), ('K', 0.4146341463414634), ('K', 0.36585365853658536), ('S', 0.0), ('F', 0.030456852791878174), ('E', 0.24742268041237114), ('T', 0.19014084507042253), ('F', 0.0), ('S', 0.07692307692307693), ('H', 0.11956521739130435), ('R', 0.25806451612903225), ('R', 0.0), ('S', 0.03076923076923077), ('C', 0.0), ('L', 0.0), ('V', 0.0), ('K', 0.05853658536585366), ('S', 0.2692307692307692), ('L', 0.31097560975609756), ('L', 0.024390243902439025), ('N', 0.4840764331210191), ('E', 0.7164948453608248), ('E', 0.22164948453608246), ('A', 0.8867924528301887), ('H', 0.1956521739130435), ('K', 0.6487804878048781), ('A', 0.8773584905660378), ('R', 0.1774193548387097), ('Y', 0.036036036036036036), ('P', 0.08088235294117647), ('P', 0.6397058823529411), ('S', 0.25384615384615383), ('P', 0.7647058823529411), ('A', 1.0)]\n"
     ]
    }
   ],
   "source": [
    "a = burial_list[0][0]\n",
    "b = burial_list[1][0]\n",
    "\n",
    "for i in range(len(seq_list)):\n",
    "    print(seq_list[i])\n",
    "    print(\"\".join([ssp_list[i][aa][1] for aa in range(len(ssp_list[i]))]))\n",
    "print()\n",
    "\n",
    "print(burial_list[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use burial to modulate similarity function in Needleman-Wunsh\n",
    "Idea: we keep the same algorithm but modify the gap or mutation cost depending of the structural information.\n",
    "\n",
    "The most easy data to use is the relative solvent accessibility (RSA) : it indicates if a residue in on the surface of the molecule or in the inside. A mutation inside of the protein is more costly than on the outside. Therefore we use the same algorythm as previopusly but multiply the similarity by a coefficient depending on the burial.\n",
    "\n",
    "\n",
    "$M_{i,j}$ is defined by recursion with the following formula:\n",
    "$$\n",
    "M_{i,j} = \\max \\begin{cases}\n",
    "  M_{i-1,j-1} + f(x) \\cdot s(a_i,b_j) \\\\\n",
    "  M_{i-1,j} - f(x) \\cdot \\text{gap cost} &\\\\\n",
    "  M_{i,j-1} - f(x) \\cdot \\text{gap cost}&\n",
    "\\end{cases}\n",
    "$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sim_burial(a, b, sim_fct=sim_blossum)->float:\n",
    "    ''' computes a similarity score ponderated by the burial score\n",
    "    a, b are tuples (str: amino-acide, float: burial score)'''\n",
    "    if a[1] == '-' or b[1] == '-':    # no burial score provided for this residue\n",
    "        coeff = 1\n",
    "    else:\n",
    "        coeff = 1.5 - (a[1] + b[1])/2 # score between 0.5 and 1.5; 0.5 means that the two residues are very exposed\n",
    "    return coeff * sim_fct(a[0], b[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment, score = progressive_alignment(burial_list, sim_burial, -10, -0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-------S-KRAREAF--------------------------------NSGKTRSLQFRIQQLEALQRMIN-------EN--------------LKS-ISGA-LASDLGKNEWTSYYEEV-AHVLEE-LDTTIKE---LPDWAED--EPVAKTRQTQQDDLYIHSEPLGVVLVIGAWNYPFNLT-I-QPMVGAVAAGNAVILK-----PSE----VSG------HMADL---LATLIPQYMDQ-NLYLVVKGGVPETTELLKERFDHIMYTGSTAVGKIVMAAAAKHLTP------VTLELG-GKSPCYVDKDC---DLDVACRRI---AWGKFMNSGQTCVAPDYILCDPSIQNQIVEK-LKKSLKDFYGEDAKQSRD--YGRIINDRHFQRVKGLI-----DNQKVAHGGTWDQSSRYIAPTIL-VDVDP--QSPVMQEEIFGPVMPIVCVRSLEEAIQFINQREKPLALY--VFSNNE---KVIKKM-IAETSSGGVTANDVIVHITVP---TL--PFGG----VGNS-GMGAYHGKKSFETFSHRRSCLVKSLLNEE---AHKARYPPSPA\n",
      "-TPNIFERMARRVREEFGRHYPLYIGGEWVDTKERMVSLNPSAPSEVVGTTAKAGKAEAEA-ALEAAWKAFKTWKDWPQEDRSRLLLKAAALMRRRKRELEAT-LVYEVGKN-WVEASADV-AEAIDF-IEYYARA---ALRYRYPAVEVVPYPGED-NESFYV---PLGAGVVIAPWNFPVAIF-T-GMIVGPVAVGNTVIAK-----PAEDAVVVGAKVFEIFHEAGFPPGVVNFLPGVGEEVGAYL-V-----EHPR-I--RF--INFTGSLEVGLKIYEAAGRLAPGQTWFKRAYVETG-GKNAIIVDETA---DFDLAAEGVVVSAYG--F-QGQKCSAASRLILTQGAYEPVLERVLKRAERLSVGP-AEENPD--LGPVVSAEQERKVLSYIEIGKNEGQLVLGGKRLEGEGYFIAPTVF-TEVPP--KARIAQEEIFGPVLSVIRVKDFAEALEVAN--DTPYGLTGGVYSRKR---EHLEWA-RREFHVGNLYFNRKITGALVG---VQ--PFGG----FKLS-GTNA----KTG-ALDY-----LRLFLEMK---AVAERF-----\n",
      "TVAAGLAVEVNQAATAAAK-----V-AR--DF--RR-----------LNNSKRASLLRTIASELEA---------------RSDDII-ARA---HLETALPEVRLTGEIAR---TANQLRLFADVVNS--GSYHQA---ILDTPNPTRAPLPKP------DIRRQQIALGPVAVFGASNFPLAFSAAGGDTASALAAGCPVIVKGHTAHPG-TSQIVAECIEQALKQEQLPQAIFTLLQGNQRALGQAL-V-----SHPE-I----KAVGFTGSVGGGRALFNLAHERPEPIPFY-G---ELG-AINPTFIFPSAMRAKADLADQFVASMTMG---C-GQFCTKPG-VVF--ALNTPETQAFIETAQSL-I---RQQSPSTLLTPGIRDSYQSQV---VSRGSDDGIDVTF-S--QAESPCVASALFVTSSENWRKHPAWEEEIFGPQSLIVVCENVADMLSLSEML-AG-SLTATIHATEEDYPQVSQLIPRLEEIAGRLVFNGWPTGVEVGYAMVHGGPYPASTHSASTSVGAEAIHRWLRPVAYQALPESLLPDSLKAENPLEIARAVDGKAA\n",
      "----------DELLEKA-K-------------KVREAW-------DVLRNATTREKNKAIKKIAEKL-------------D------------ERRKEILEANRIDVEKARERGVKESLVDRLALNDKRIDEXIKACETVIGLKDPVGEVIDSWVREDGLRIARVRVPIGPIGII----YESRPNVTVETTILALKSGNTILLRGGSDALNSNKAIVSA-IREALKETEIPESSVEFIENTDRSL---VLEXIRLREYLSLVIPR-----------GGYGLISFV--RDNA-T--V-PVLETGVGNCHIFVDESA-----DLK-KAVPVIINAKTQRPG-TCNAAEKLL--------VHEKIAKEFL---------------------P-----VI--VEELRKHGVEVR--G-CEKTRE-IVPDVVPATEDDW---PT---EYLDLIIAIKVVKNVDEAIEHIKKYSTGHS--ESILT--ENYSNAKKFVS--EIDAAAVYVN-------ASTRFTDGGQF-G----FGAEIG---IS---T-QRFHARGPVGLRE-LTTYK-FVVLGEYHV-RE\n"
     ]
    }
   ],
   "source": [
    "seq_list = []\n",
    "for seq in alignment:\n",
    "    for i in range(len(seq)):\n",
    "        if seq[i] == '-':\n",
    "            seq[i] = ('-', '-')\n",
    "    seq_list.append(\"\".join([aa[0] for aa in seq]))\n",
    "\n",
    "for seq in seq_list:\n",
    "    print(seq)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run simulations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### With affine gap but without PDB\n",
    "(takes about 3min30 for 5 alignments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# train_data = 5\n",
    "\n",
    "# for i in range(1, train_data+1):\n",
    "#     data_file = f\"data/balibase/RV11.unaligned/BB11{str(i).zfill(3)}.fasta\"\n",
    "#     seqs, ids = read_fasta(data_file)\n",
    "#     for gap_opening in [-5, -8,-10,-12,-15, -20]:\n",
    "#         for gap_extension in [-0.1,-0.5,-1,-2]:\n",
    "#             result_file = f\"data/results/affineWithoutPDB/results{-gap_opening}_{-gap_extension}/BB11{str(i).zfill(3)}.fasta\"\n",
    "#             alignment, score = progressive_alignment(seqs, sim_blossum, gap_opening, gap_extension)\n",
    "#             write_fasta(result_file, alignment, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### With affine gap and PDB\n",
    "(takes about 5min for 5 alignments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 1452.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 1457.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 1462.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 1525.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 1571.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 1578.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3536.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3752.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2811.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 2890.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 2990.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 3000.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue SER het=  resseq=283 icode= > not found in DSSP for 1ov3_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9025.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9069.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 9113.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9240.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8685.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8701.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8733.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 9485.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15917.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15948.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15979.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16010.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 16041.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 16142.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 16238.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16349.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue MSE het=H_MSE resseq=78 icode= > not found in DSSP for 1o20_A\n",
      "residue <Residue MSE het=H_MSE resseq=189 icode= > not found in DSSP for 1o20_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6519.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6569.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7067.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7068.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7077.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7277.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 8147.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 8164.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15828.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15863.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15879.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 15904.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7199.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7214.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7229.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7284.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6929.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6990.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue LLP het=H_LLP resseq=253 icode= > not found in DSSP for 1bw0_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6283.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6298.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6313.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6348.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7196.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7214.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7230.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7395.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7029.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7045.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7060.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7240.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15495.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15518.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15537.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 15560.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15579.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15746.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15893.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16040.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6341.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6361.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 6381.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 6502.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue MSE het=H_MSE resseq=112 icode= > not found in DSSP for 1fg3_A\n",
      "residue <Residue MSE het=H_MSE resseq=237 icode= > not found in DSSP for 1fg3_A\n",
      "residue <Residue MSE het=H_MSE resseq=265 icode= > not found in DSSP for 1fg3_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11508.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11523.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 11538.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 11553.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11568.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11622.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 11656.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 11676.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue MSE het=H_MSE resseq=92 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=93 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=95 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=239 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=265 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=272 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=288 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue MSE het=H_MSE resseq=326 icode= > not found in DSSP for 1h1c_A\n",
      "residue <Residue LYS het=  resseq=335 icode= > not found in DSSP for 1h1c_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11404.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11406.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 11408.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 11410.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 11412.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 11641.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 11852.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 12049.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue LLP het=H_LLP resseq=203 icode= > not found in DSSP for 1jg8_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15691.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15692.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 15693.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 15694.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 15695.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 15920.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain C is discontinuous at line 16160.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain D is discontinuous at line 16421.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "residue <Residue LLP het=H_LLP resseq=266 icode= > not found in DSSP for 1ax4_A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 7091.\n",
      "  warnings.warn(\n",
      "/home/flopau/Documents/.venv/lib/python3.8/site-packages/Bio/PDB/StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 7264.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "train_data = 5\n",
    "\n",
    "for i in range(1, train_data+1):\n",
    "    data_file = f\"data/balibase/RV11.unaligned/BB11{str(i).zfill(3)}.fasta\"\n",
    "    seqs, ssps, ids = read_and_align_fasta_and_SSP(data_file)\n",
    "    burial_list = [ssp2tupple(ssp) for ssp in ssps]\n",
    "    for gap_opening in [-18, -20]:\n",
    "        for gap_extension in [-1,-2,-4, -5]:\n",
    "            result_file = f\"data/results/affineWithPDB2/results{-gap_opening}_{-gap_extension}/BB11{str(i).zfill(3)}.fasta\"\n",
    "            alignment, score = progressive_alignment(burial_list, sim_burial, gap_opening, gap_extension)\n",
    "            alignment = [\"\".join([aa[0] for aa in seq]) for seq in alignment]\n",
    "            write_fasta(result_file, alignment, ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### With ClustalW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "\n",
    "# train_data = 38\n",
    "\n",
    "# for i in range(1, train_data+1):\n",
    "#     infile = f\"data/balibase/RV11.unaligned/BB11{str(i).zfill(3)}.fasta\"\n",
    "#     outfile = f\"data/balibase/results/clustalo/BB11{str(i).zfill(3)}.fasta\"\n",
    "#     cli = ClustalOmegaCommandline( \"clustalo\", infile=infile, outfile=outfile )\n",
    "#     try:\n",
    "#         stdout , stderr = cli()\n",
    "#     except Exception as e:\n",
    "#         print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Benchmark\n",
    "For the sequences from balivase, we want to compare:\n",
    "- our computed alignments (progressive alignment using affine/linear gap with/without structural information)\n",
    "- the prediction from clustalW (best alignment without structural info)\n",
    "- the consensus alignment given in the fasta file\n",
    "\n",
    "`balibase2.py` contains a function `compare_seq` which takes in input a reference and a test alignment (in form of a dictionnary of sequences and its list of keys), and return two scores:\n",
    "- the sum-of-pairs score (SPS): measures if *some* sequences are correctly aligned\n",
    "- the column score: measures if *every* sequences are correctly aligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results of clustal omega compared to balibase reference:\n",
      "____________________________________________________\n",
      "    id        SP    TC\n",
      "0    1  0.901316  0.80\n",
      "1    2  0.483283  0.00\n",
      "2    3  0.604775  0.42\n",
      "3    4  0.522447  0.38\n",
      "4    5  0.400970  0.10\n",
      "5    6  0.382857  0.20\n",
      "6    7  0.607697  0.33\n",
      "7    8  0.379630  0.18\n",
      "8    9  0.324786  0.00\n",
      "9   10  0.320652  0.17\n",
      "10  11  0.217188  0.09\n",
      "11  12  0.862975  0.76\n",
      "12  13  0.053061  0.00\n",
      "13  14  0.766741  0.62\n",
      "14  15  0.719833  0.56\n",
      "15  16  0.574000  0.32\n",
      "16  17  0.768152  0.67\n",
      "17  18  0.501323  0.25\n",
      "18  19  0.574087  0.16\n",
      "19  20  0.593892  0.19\n",
      "20  21  0.493750  0.35\n",
      "21  22  0.152047  0.00\n",
      "22  23  0.530314  0.23\n",
      "23  24  0.236842  0.00\n",
      "24  25  0.104575  0.00\n",
      "25  26  0.125908  0.00\n",
      "26  27  0.355337  0.14\n",
      "27  28  0.583053  0.00\n",
      "28  29  0.479798  0.39\n",
      "29  30  0.579601  0.21\n",
      "30  31  0.474688  0.13\n",
      "31  32  0.613805  0.36\n",
      "32  33  0.226607  0.00\n",
      "33  34  0.557555  0.22\n",
      "34  35  0.415152  0.12\n",
      "35  36  0.558796  0.23\n",
      "36  37  0.480212  0.27\n",
      "37  38  0.747050  0.55\n"
     ]
    }
   ],
   "source": [
    "# comparing clustal omega to balibase reference\n",
    "train_data = 38\n",
    "\n",
    "Res = {'id':[], 'SP':[], 'TC':[]}\n",
    "\n",
    "for i in range(1, train_data+1):\n",
    "    file_bali = f\"data/balibase/RV11.aligned/BB11{str(i).zfill(3)}.fasta\"\n",
    "    file_clustal = f\"data/results/clustalo/BB11{str(i).zfill(3)}.fasta\"\n",
    "\n",
    "    refkeylist, refseq = read_fasta2dict(file_bali)\n",
    "    testkeylist, testseq = read_fasta2dict(file_clustal)\n",
    "\n",
    "    sp, tc = compare_seq(refseq, refkeylist, testseq, testkeylist)\n",
    "    Res['id'].append(i)\n",
    "    Res['SP'].append(sp)\n",
    "    Res['TC'].append(tc)\n",
    "\n",
    "df_clustalo = pd.DataFrame(Res)\n",
    "print(\"Results of clustal omega compared to balibase reference:\")\n",
    "print('____________________________________________________')\n",
    "print(df_clustalo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# comparing our results to culstalo and balibase reference\n",
    "train_data = 5\n",
    "\n",
    "Res = {'id':[], 'ref':[], \"pdb\":[], 'gap_opening':[], 'gap_extension':[], 'SP':[], 'TC':[]}\n",
    "\n",
    "\n",
    "for i in range(1, train_data+1):\n",
    "    ref_file1 = f\"data/balibase/RV11.aligned/BB11{str(i).zfill(3)}.fasta\"\n",
    "    ref_file2 = f\"data/results/clustalo/BB11{str(i).zfill(3)}.fasta\"\n",
    "\n",
    "    refkeylist1, refseq1 = read_fasta2dict(ref_file1)\n",
    "    refkeylist2, refseq2 = read_fasta2dict(ref_file2)\n",
    "\n",
    "    for gap_opening in [-8, -10, -12, -15]:\n",
    "        for gap_extension in [-2, -1, -0.5]:\n",
    "            test_file = f\"data/results/affineWithoutPDB/results{-gap_opening}_{-gap_extension}/BB11{str(i).zfill(3)}.fasta\"\n",
    "            testkeylist, testseq = read_fasta2dict(test_file)\n",
    "\n",
    "            sp1, tc1 = compare_seq(refseq1, refkeylist1, testseq, testkeylist)\n",
    "            sp2, tc2 = compare_seq(refseq2, refkeylist2, testseq, testkeylist)\n",
    "            Res['id'].append(i)\n",
    "            Res['ref'].append('balibase')\n",
    "            Res['gap_opening'].append(gap_opening)\n",
    "            Res['gap_extension'].append(gap_extension)\n",
    "            Res['SP'].append(sp1)\n",
    "            Res['TC'].append(tc1)\n",
    "            Res['pdb'].append(False)\n",
    "\n",
    "            Res['id'].append(i)\n",
    "            Res['ref'].append('clustalo')\n",
    "            Res['gap_opening'].append(gap_opening)\n",
    "            Res['gap_extension'].append(gap_extension)\n",
    "            Res['SP'].append(sp2)\n",
    "            Res['TC'].append(tc2)\n",
    "            Res['pdb'].append(False)\n",
    "\n",
    "            test_file = f\"data/results/affineWithPDB2/results{-gap_opening}_{-gap_extension}/BB11{str(i).zfill(3)}.fasta\"\n",
    "            testkeylist, testseq = read_fasta2dict(test_file)\n",
    "\n",
    "            sp1, tc1 = compare_seq(refseq1, refkeylist1, testseq, testkeylist)\n",
    "            sp2, tc2 = compare_seq(refseq2, refkeylist2, testseq, testkeylist)\n",
    "            Res['id'].append(i)\n",
    "            Res['ref'].append('balibase')\n",
    "            Res['gap_opening'].append(gap_opening)\n",
    "            Res['gap_extension'].append(gap_extension)\n",
    "            Res['SP'].append(sp1)\n",
    "            Res['TC'].append(tc1)\n",
    "            Res['pdb'].append(True)\n",
    "\n",
    "            Res['id'].append(i)\n",
    "            Res['ref'].append('clustalo')\n",
    "            Res['gap_opening'].append(gap_opening)\n",
    "            Res['gap_extension'].append(gap_extension)\n",
    "            Res['SP'].append(sp2)\n",
    "            Res['TC'].append(tc2)\n",
    "            Res['pdb'].append(True)\n",
    "\n",
    "df = pd.DataFrame(Res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     mol       ref gap_opening gap_extension        SP    TC    PDB\n",
      "0      4  clustalo          15             1  0.015670  0.00   True\n",
      "1      4  balibase          15             1  0.006734  0.00   True\n",
      "2      5  clustalo          15             1  0.187913  0.00   True\n",
      "3      5  balibase          15             1  0.159937  0.00   True\n",
      "4      2  clustalo          15             1  0.000000  0.00   True\n",
      "..   ...       ...         ...           ...       ...   ...    ...\n",
      "435    2  balibase          12           0.1  0.117021  0.00  False\n",
      "436    1  clustalo          12           0.1  0.655702  0.48  False\n",
      "437    1  balibase          12           0.1  0.719298  0.60  False\n",
      "438    3  clustalo          12           0.1  0.323699  0.08  False\n",
      "439    3  balibase          12           0.1  0.263926  0.07  False\n",
      "\n",
      "[440 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "# comparing our results to culstalo and balibase reference\n",
    "train_data = 5\n",
    "\n",
    "# Res = {'id':[], 'ref':[], \"pdb\":[], 'gap_opening':[], 'gap_extension':[], 'SP':[], 'TC':[], 'pdb':[]}\n",
    "Res = []\n",
    "\n",
    "for root, dirs, files in os.walk(\"data/results/affineWithPDB2\"):\n",
    "    for file in files:\n",
    "        folder = root.split('/')[-1].lstrip('results')\n",
    "        gap_opening, gap_extension = folder.split('_')\n",
    "        ref_file_bali = f\"data/balibase/RV11.aligned/{file}\"\n",
    "        ref_file_clustalo = f\"data/results/clustalo/{file}\"\n",
    "        refkeylist_bali, refseq_bali = read_fasta2dict(ref_file_bali)\n",
    "        refkeylist_clustalo, refseq_clustalo = read_fasta2dict(ref_file_clustalo)\n",
    "        test_file = f\"{root}/{file}\"\n",
    "        testkeylist, testseq = read_fasta2dict(test_file)\n",
    "\n",
    "        sp_bali, tc_bali = compare_seq(refseq_bali, refkeylist_clustalo, testseq, testkeylist)\n",
    "        sp_clustalo, tc_clustalo = compare_seq(refseq_clustalo, refkeylist_clustalo, testseq, testkeylist)\n",
    "\n",
    "        Res.append([int(file.split('.')[0][-3:]),\"clustalo\",gap_opening,gap_extension,sp_clustalo,tc_clustalo,True])\n",
    "        Res.append([int(file.split('.')[0][-3:]),\"balibase\",gap_opening,gap_extension,sp_bali,tc_bali,True])\n",
    "\n",
    "for root, dirs, files in os.walk(\"data/results/affineWithoutPDB\"):\n",
    "    for file in files:\n",
    "        folder = root.split('/')[-1].lstrip('results')\n",
    "        gap_opening, gap_extension = folder.split('_')\n",
    "        ref_file_bali = f\"data/balibase/RV11.aligned/{file}\"\n",
    "        ref_file_clustalo = f\"data/results/clustalo/{file}\"\n",
    "        refkeylist_bali, refseq_bali = read_fasta2dict(ref_file_bali)\n",
    "        refkeylist_clustalo, refseq_clustalo = read_fasta2dict(ref_file_clustalo)\n",
    "        test_file = f\"{root}/{file}\"\n",
    "        testkeylist, testseq = read_fasta2dict(test_file)\n",
    "\n",
    "        sp_bali, tc_bali = compare_seq(refseq_bali, refkeylist_clustalo, testseq, testkeylist)\n",
    "        sp_clustalo, tc_clustalo = compare_seq(refseq_clustalo, refkeylist_clustalo, testseq, testkeylist)\n",
    "\n",
    "        Res.append([int(file.split('.')[0][-3:]),\"clustalo\",gap_opening,gap_extension,sp_clustalo,tc_clustalo,False])\n",
    "        Res.append([int(file.split('.')[0][-3:]),\"balibase\",gap_opening,gap_extension,sp_bali,tc_bali,False])\n",
    "\n",
    "df = pd.DataFrame(Res, columns =['mol', 'ref', 'gap_opening', 'gap_extension', 'SP', 'TC', 'PDB'])\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mol</th>\n",
       "      <th>SP</th>\n",
       "      <th>TC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <th>gap_opening</th>\n",
       "      <th>gap_extension</th>\n",
       "      <th>PDB</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">clustalo</th>\n",
       "      <th>10</th>\n",
       "      <th>1</th>\n",
       "      <th>False</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.487251</td>\n",
       "      <td>0.252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <th>1</th>\n",
       "      <th>False</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.482489</td>\n",
       "      <td>0.246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>balibase</th>\n",
       "      <th>10</th>\n",
       "      <th>1</th>\n",
       "      <th>False</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.480398</td>\n",
       "      <td>0.276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">clustalo</th>\n",
       "      <th>12</th>\n",
       "      <th>0.5</th>\n",
       "      <th>False</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.477708</td>\n",
       "      <td>0.228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <th>1</th>\n",
       "      <th>False</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.460479</td>\n",
       "      <td>0.210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>balibase</th>\n",
       "      <th>15</th>\n",
       "      <th>0.5</th>\n",
       "      <th>True</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.097811</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clustalo</th>\n",
       "      <th>12</th>\n",
       "      <th>0.5</th>\n",
       "      <th>True</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.097159</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">balibase</th>\n",
       "      <th>12</th>\n",
       "      <th>0.5</th>\n",
       "      <th>True</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.092671</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <th>0.5</th>\n",
       "      <th>True</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.087103</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clustalo</th>\n",
       "      <th>8</th>\n",
       "      <th>0.5</th>\n",
       "      <th>True</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.084853</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          mol        SP     TC\n",
       "ref      gap_opening gap_extension PDB                        \n",
       "clustalo 10          1             False  3.0  0.487251  0.252\n",
       "         15          1             False  3.0  0.482489  0.246\n",
       "balibase 10          1             False  3.0  0.480398  0.276\n",
       "clustalo 12          0.5           False  3.0  0.477708  0.228\n",
       "         8           1             False  3.0  0.460479  0.210\n",
       "...                                       ...       ...    ...\n",
       "balibase 15          0.5           True   3.0  0.097811  0.000\n",
       "clustalo 12          0.5           True   3.0  0.097159  0.000\n",
       "balibase 12          0.5           True   3.0  0.092671  0.000\n",
       "         8           0.5           True   3.0  0.087103  0.000\n",
       "clustalo 8           0.5           True   3.0  0.084853  0.000\n",
       "\n",
       "[88 rows x 3 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['ref', 'gap_opening', 'gap_extension', 'PDB']).mean().sort_values(by='SP', ascending=False)\n",
    "# df[df.ref=='clustalo'].groupby(['gap_opening', 'gap_extension']).mean().sort_values(ascending=False)\n",
    "# df[df.ref=='balibase'].groupby(['gap_opening', 'gap_extension'])['SP'].mean().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When sorting the results by sum of pairs score, we find that the best parameters seem to be `gap_opening` = -12 and `gap_extension` = -1.\n",
    "\n",
    "With PDB performs very badly. For better performance, we have to improve `sim_burial`."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
